---
title: "all_data_results"
author: "Sejun Kim"
date: '2022 2 18 '
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(Rsamtools)
library(pheatmap)
library(DESeq2)
library(RColorBrewer)
library(factoextra)
library(ggfortify)
library(ggrepel)
library(Homo.sapiens)
library(VennDiagram)

figure <- 'd:/Projects/UCPH/Work/master_thesis/figures/'
dataset <- 'd:/Projects/UCPH/Work/master_thesis/datasets/'
atac <- 'd:/Projects/UCPH/Work/atacseq/'
peaks <- paste0(atac,'new/')
motifs <- paste0(peaks, 'motifs/')
rna <- 'd:/Projects/UCPH/Work/rnaseq/'
figures <- 'd:/Projects/UCPH/Work/master_thesis/figures_wo_legend/'

```


# atacseq analysis

## Peak-annotated gene orgarnizing from GREAT


```{r}
count <- read.table('d:/Projects/UCPH/Work/atacseq/new/raw_count_new.txt') %>%
  rownames_to_column('merged')

colnames(count) <- c('merged','si673_12h_1','si673_12h_2','si673_12h_3',
                                   'si673_24h_1', 'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 'si673_4h_3', 
                     'siNT_1', 'siNT_2', 'siNT_3')

anno <- read.delim('d:/Projects/UCPH/Work/atacseq/new/peak_annotation.txt', comment.char = '#', 
                   header = F) %>%
  mutate(merged = V1, 
         gene1 = str_split(str_split(.$V2, ', ', simplify = T)[,1],
                         '\\(', simplify = T)[,1], 
         gene2 = str_split(str_split(.$V2, ', ', simplify = T)[,2],
                         '\\(', simplify = T)[,1], 
         gene3 = str_split(str_split(.$V2, ', ', simplify = T)[,3],
                         '\\(', simplify = T)[,1], 
         gene4 = str_split(str_split(.$V2, ', ', simplify = T)[,4],
                         '\\(', simplify = T)[,1], 
         gene5 = str_split(str_split(.$V2, ', ', simplify = T)[,5],
                         '\\(', simplify = T)[,1], 
         gene6 = str_split(str_split(.$V2, ', ', simplify = T)[,6],
                         '\\(', simplify = T)[,1], 
         gene7 = str_split(str_split(.$V2, ', ', simplify = T)[,7],
                         '\\(', simplify = T)[,1]) %>%
  dplyr::select(-c(V1, V2)) 




anno1 <- anno %>% filter(gene1!="") %>% dplyr::select(merged, genes=gene1)
anno2 <- anno %>% filter(gene2!="") %>% dplyr::select(merged, genes=gene2)
anno3 <- anno %>% filter(gene3!="") %>% dplyr::select(merged, genes=gene3)
anno4 <- anno %>% filter(gene4!="") %>% dplyr::select(merged, genes=gene4)
anno5 <- anno %>% filter(gene5!="") %>% dplyr::select(merged, genes=gene5)
anno6 <- anno %>% filter(gene6!="") %>% dplyr::select(merged, genes=gene6)
anno7 <- anno %>% filter(gene7!="") %>% dplyr::select(merged, genes=gene7)

annot <- bind_rows(anno1, anno2, anno3, anno4, anno5, anno6, anno7) %>% arrange(merged)
write.table(annot,paste0(dataset, 'atac_peak_annotation_2.txt'), row.names = F)

annotation <- annot %>%
  mutate(genes = str_split(.$genes, ' ', simplify = T)[,1])


total <- full_join(annotation, count, by = 'merged' ) %>%
  na.omit() 

total_count <- as.matrix(total[3:ncol(total)])
rownames(total_count) <-total$merged


# design
design <- data.frame(treatment = c( 'si673_12h','si673_12h','si673_12h',
                                   'si673_24h', 'si673_24h','si673_24h',
                                   'si673_4h', 'si673_4h', 'si673_4h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)'))

rownames(design) <- colnames(total)[3:length(total)]



# delete outlier
ct_vs_12 <- cbind(total_count[,1:3], total_count[,10:12])

ct_vs_12_3_delete <- ct_vs_12 %>% as.data.frame() %>% dplyr::select(-si673_12h_3) %>% as.matrix()



ct_vs_12_3_delete_design <- c('si673_12h', 'si673_12h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)') %>% as.data.frame()
rownames(ct_vs_12_3_delete_design) <- c('si673_12h_1','si673_12h_2','siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_12_3_delete_design) <- 'treatment'


```


# atac Deseq analysis

```{r}
# DESeq2

dds <- DESeqDataSetFromMatrix(countData = total_count,
                               colData = design,
                               design = ~treatment)
deseq <- DESeq(dds)

# sort result with logFC threshold 0, padj < 0.1
result <- results(deseq)

# confirm result
summary(result)



# make tibble from DESeq result
result_tb <- result %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
write.csv(result_tb,paste0(dataset, 'atac_deseq_total_result.csv'), row.names = F)
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd <- vst(dds) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists <- dist( t( assay(vsd) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap <- pheatmap(sampleDistMatrix, 
                        # main = "Sample distance clustering",
                        annotation_col = design, 
                        annotation_legend = F)
distheatmap
ggsave(distheatmap, file = paste0(figures, 'atac_total_sample_dist_heatmap.png'))

```



# PCA

```{r}
pca <- prcomp(t(assay(vsd)), center=TRUE, scale.=FALSE)


pca.x <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var <- summary(pca)$importance['Proportion of Variance',] * 100

pca_2 <- ggplot(pca.x, aes(x = PC1, y = PC2, color = design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2

ggsave(pca_2, file = paste0(figures,'atac_total_pca_plot.png'))
```


# Deseq analysis - separately 4h

```{r}
# sort result with logFC threshold 0, padj < 0.1
result_4 <- results(deseq, contrast = c('treatment', 'si673_4h', 'siNT(Ctrl)'))

# confirm result
summary(result_4)

# make tibble from DESeq result
result_tb_4 <- result_4 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot - 4h

```{r}
# make a plot - volcano plot
result_padj_4 <- result_tb_4 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 


subset_4 <- result_padj_4 %>%
  filter(padj < 0.1)

volcano_4 <- ggplot(result_tb_4, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.1') +
  geom_text_repel(aes(label = genes), data = subset_4, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_4

ggsave(volcano_4, file = paste0(figures,'volcano_plot_atac_4.png'))
```



# Deseq analysis - 12h

```{r}
# DESeq2

result_12 <- results(deseq, contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))


# confirm result
summary(result_12)

# make tibble from DESeq result
result_tb_12 <- result_12 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot - 12h

```{r}
# make a plot - volcano plot
result_padj_12 <- result_tb_12 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 



subset_12 <- result_padj_12 %>%
  filter(padj < 0.1)

volcano_12 <- ggplot(result_tb_12, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.1') +
  geom_text_repel(aes(label = genes), data = subset_12, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_12

ggsave(volcano_12, file = paste0(figures,'volcano_plot_atac_12.png'))
```





# Data transfrom - 12h

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_12 <- assay(vsd) %>% as.data.frame() %>% 
  dplyr::select(c('si673_12h_1', 'si673_12h_2', 'si673_12h_3', 
                                         'siNT_1','siNT_2','siNT_3')) %>%
  as.matrix()

```



# PCA - 12h

```{r}

pca_12 <- prcomp(t(vsd_12), center=TRUE, scale.=FALSE)
fviz_eig(pca_12) # elbow method

pca.x_12 <- pca_12$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_12 <- summary(pca_12)$importance['Proportion of Variance',] * 100

pca_2_12 <- ggplot(pca.x_12, aes(x = PC1, y = PC2, color = treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_12[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_12[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_12

ggsave(pca_2_12, file = paste0(figures,'pca_plot_12h.png'))
```







# Deseq analysis - 24h

```{r}
# DESeq2
result_24 <- results(deseq, lfcThreshold = 0.25, alpha = 0.05,
                     contrast = c('treatment', 'si673_24h', 'siNT(Ctrl)'))



# confirm result
summary(result_24)

# make tibble from DESeq result
result_tb_24 <- result_24 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_24, file = paste0(dataset,'result_atac_24.csv'))
```



# volcano plot - 24h

```{r}
# make a plot - volcano plot
result_padj_24 <- result_tb_24 %>%
  arrange(padj) %>%
  filter(padj < 0.05) 



subset_24 <- result_padj_24 %>%
  filter(padj < 0.05)

volcano_24 <- ggplot(result_tb_24, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  geom_text_repel(aes(label = genes), data = subset_24, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_24

ggsave(volcano_24, file = paste0(figures,'volcano_plot_atac_24.png'))
```




# Data transfrom - 24h

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_24 <- assay(vsd) %>% as.data.frame() %>% 
  dplyr::select(c('si673_24h_1', 'si673_24h_2', 'si673_24h_3', 
                                         'siNT_1','siNT_2','siNT_3')) %>%
  as.matrix()
```



# PCA - 24h

```{r}

pca_24 <- prcomp(t(vsd_24), center=TRUE, scale.=FALSE)
fviz_eig(pca_24) # elbow method

pca.x_24 <- pca_24$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_24 <- summary(pca_24)$importance['Proportion of Variance',] * 100

pca_2_24 <- ggplot(pca.x_24, aes(x = PC1, y = PC2, color = treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_24[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_24[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_24

ggsave(pca_2_24, file = paste0(figures,'pca_plot_24h.png'))
```



# Deseq analysis - 12h_delete

```{r}
# DESeq2

dds_12d <- DESeqDataSetFromMatrix(countData = ct_vs_12_3_delete,
                               colData = ct_vs_12_3_delete_design,
                               design = ~treatment)
deseq_12d <- DESeq(dds_12d)

# sort result with logFC threshold 0.25, padj < 0.05
result_12d <- results(deseq_12d, lfcThreshold = 0.25, alpha = 0.05, 
                      contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))

# confirm result
summary(result_12d)

# make tibble from DESeq result
result_tb_12d <- result_12d %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_12d, file = paste0(dataset,'result_atac_12h_delete.csv'))
```



# volcano plot - 12h_delete

```{r}
result_tb_12d <- read.csv(paste0(dataset,'result_atac_12h_delete.csv')) 

# make a plot - volcano plot
result_padj_12d <- result_tb_12d %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

write_csv(result_padj_12d, file = paste0(dataset,'high_padj_atac_12_delete.csv'))

subset_12d <- result_padj_12d %>%
  filter(padj < 0.03, abs(log2FoldChange) > 1.5)

volcano_12d <- ggplot(result_tb_12d, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  # geom_text_repel(aes(label = genes), data = subset_12d, max.overlaps = 1000) +
  theme_bw()
# +
#   theme(legend.position = 'none')
volcano_12d

ggsave(volcano_12d, file = paste0(figure,'volcano_plot_atac_12_delete.png'))
```




# Data transfrom - 12h_delete

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_12d <- vst(dds_12d) # variance stabilizing transformation

```

# heatmap sample dist - 12h_delete

```{r}

sampleDists_12d <- dist( t( assay(vsd_12d) ) )
sampleDistMatrix_12d <- as.matrix( sampleDists_12d ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_12d <- pheatmap(sampleDistMatrix_12d
                        # main = "Sample distance clustering"
                        )
distheatmap_12d
ggsave(distheatmap_12d, file = paste0(figure,'sample_dist_heatmap_12h_delete.png'))

```


# sample correlation plot - 12h_delete

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_12d <- cor(assay(vsd_12d))
# generate heatmap
samplecorheatmap_12d <- pheatmap(
  as.matrix(sampleCor_12d),
  # main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_12d
ggsave(samplecorheatmap_12d, file = paste0(figure,'sample_correlation_heatmap_12h_delete.png'))

```


# PCA - 12h_delete

```{r}

pca_12d <- prcomp(t(assay(vsd_12d)), center=TRUE, scale.=FALSE)
fviz_eig(pca_12d) # elbow method

pca.x_12d <- pca_12d$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_12d <- summary(pca_12d)$importance['Proportion of Variance',] * 100

pca_2_12d <- ggplot(pca.x_12d, aes(x = PC1, y = PC2, color = ct_vs_12_3_delete_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_12d[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_12d[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_12d

ggsave(pca_2_12d, file = paste0(figures,'pca_plot_12h_delete.png'))
```


# heatmap of genes for higher padj values - 12h_delete


```{r, fig.height = 11.5, fig.width = 10}
x_12d<- assay(vsd_12d)
rownames(x_12d) <- result_tb_12d$genes
mat_12d <- x_12d[head(order(result_tb_12d$padj),50), ]

mat_12d <- mat_12d - rowMeans(mat_12d)
adjheatmap_12d <- pheatmap(mat_12d, 
                           annotation_col = ct_vs_12_3_delete_design, 
                           annotation_legend = F,
                           cellheight = 10)
adjheatmap_12d
ggsave(adjheatmap_12d, file = paste0(figures,'gene_significant_heatmap_12h_delete.png'))

```


```{r}
plotDispEsts(deseq_12d)
```



# MA plot - 12h_delete

```{r}
# make a plot - MA plot
ggplot(result_tb_12d, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```


# motifs

```{r}
prefix <- 'knownResults_'
motif_nt <- read_tsv(paste0(motifs, prefix, 'siNT.txt')) %>%
  mutate(TFs = str_split(str_split(.$`Motif Name`, '/', simplify = T)[,1],
                         '\\(', simplify = T)[,1],
         Genes = str_split(str_split(str_split(.$`Motif Name`, '/', simplify = T)[,1],
                         '\\(', simplify = T)[,2],'\\)', simplify = T)[,1],
         from = str_split(.$`Motif Name`, '/', simplify = T)[,2], 
         Target_percent = as.numeric(str_split(.$`% of Target Sequences with Motif`,
                                               '\\%', simplify = T)[,1]), 
         Bg_percent = as.numeric(str_split(.$`% of Background Sequences with Motif`,
                                           '\\%', simplify = T)[,1]), 
         log_pvalue = `Log P-value`) %>%
  dplyr::select(c(TFs, Genes, from, log_pvalue, Target_percent,Bg_percent))%>%
  filter(exp(log_pvalue) <=1e-10, Target_percent - Bg_percent >= 3) %>%
  mutate(group = 'siNT')
  


motif_12d <- read_tsv(paste0(motifs, prefix, 'si673_12h_d.txt')) %>%
  mutate(TFs = str_split(str_split(.$`Motif Name`, '/', simplify = T)[,1],
                         '\\(', simplify = T)[,1],
         Genes = str_split(str_split(str_split(.$`Motif Name`, '/', simplify = T)[,1],
                         '\\(', simplify = T)[,2],'\\)', simplify = T)[,1],
         from = str_split(.$`Motif Name`, '/', simplify = T)[,2], 
         Target_percent = as.numeric(str_split(.$`% of Target Sequences with Motif`,
                                               '\\%', simplify = T)[,1]), 
         Bg_percent = as.numeric(str_split(.$`% of Background Sequences with Motif`,
                                           '\\%', simplify = T)[,1]), 
         log_pvalue = `Log P-value`) %>%
  dplyr::select(c(TFs, Genes, from, log_pvalue, Target_percent,Bg_percent))%>%
  filter(exp(log_pvalue) <=1e-10, Target_percent - Bg_percent >= 3)%>%
  mutate(group = 'si673_12h')




write.table(motif_nt, paste0(dataset, 'organized_motif_siNT.csv'))
write.table(motif_12d, paste0(dataset, 'organized_motif_si673_12h_delete.csv'))

```


# comparing motifs in 12h deseq

```{r}
atacseq_12d <- read_csv(paste0(dataset,'result_atac_12h_delete.csv'))%>%
  na.omit() %>%
  dplyr::select(c(TFs = genes, log2FoldChange, padj))

atac_12d_motifs <- atacseq_12d  %>%
  inner_join(., motif_12d, by = 'TFs') %>%
  arrange(padj) %>%
  filter(!(TFs %in% motif_nt$TFs))%>%
  filter(padj < 0.05)

write.csv(atac_12d_motifs, paste0(dataset, "atac_motif_significant_12_delete.csv"))

```


## motif analaysis with rnaseq


```{r}

atac_deseq_12d <- read_csv('d:/Projects/UCPH/Work/master_thesis/datasets/result_atac_12h_delete.csv') %>%
  dplyr::select(c(merged, log2FoldChange, padj, TFs=genes))

rnaseq <- read_csv(paste0(dataset, 'rnaseq_deseq_final_result_total.csv')) %>%
  dplyr::select(c(gene, log2FoldChange, padj, TFs=symbol))

total_motifs <- read_tsv(paste0(dataset, 'atac_12h_motifs_total_annotates.txt')) 

col <- colnames(total_motifs)

col[1] <- 'PeakID'
col[22] <- 'KLF3'
col[23] <- 'KLF14'
col[24] <- 'MYB'
colnames(total_motifs) <- col

klf3 <- total_motifs %>%
  filter(!is.na(KLF3)) %>%
  dplyr::select(c(TFs = `Gene Name`, `Gene Alias`, `Gene Description`, KLF3))

klf14 <- total_motifs %>%
  filter(!is.na(KLF14)) %>%
  dplyr::select(c(TFs = `Gene Name`, `Gene Alias`, `Gene Description`, KLF14))

myb <- total_motifs %>%
  filter(!is.na(MYB)) %>%
  dplyr::select(c(TFs = `Gene Name`, `Gene Alias`, `Gene Description`, MYB))


motif_12d <- atac_deseq_12d %>%
  na.omit() %>%
  mutate(klf3 = ifelse(TFs %in% klf3$TFs, 'klf3_related', 'klf3_not'),
         klf14 = ifelse(TFs %in% klf14$TFs, 'klf14_related', 'klf14_not'),
         myb = ifelse(TFs %in% myb$TFs, 'myb_related', 'myb_not'),
         DEgenes = ifelse(padj <= 0.05, 'significant', 'not_significant'))


atac_klf3_table <- table('DE gene' = motif_12d$DEgenes, 'klf3 gene' = motif_12d$klf3)
atac_klf3_table

atac_klf3_test <-fisher.test(atac_klf3_table)
atac_klf3_test$estimate
atac_klf3_test$p.value

atac_klf14_table <- table('DE gene' = motif_12d$DEgenes, 'klf14 gene' = motif_12d$klf14)
atac_klf14_table

atac_klf14_test <-fisher.test(atac_klf14_table)
atac_klf14_test$estimate
atac_klf14_test$p.value


atac_myb_table <- table('DE gene' = motif_12d$DEgenes, 'myb gene' = motif_12d$myb)
atac_myb_table

atac_myb_test <-fisher.test(atac_myb_table)
atac_myb_test$estimate
atac_myb_test$p.value

klf_12d_sig <- motif_12d %>% filter(padj < 0.05, (klf3 == 'klf3_related'| klf14 == 'klf14_related')) %>%
  dplyr::select(-myb) %>%
  arrange(log2FoldChange)

klf3_atac_sig <- klf_12d_sig %>% filter(klf3 == 'klf3_related')
klf14_atac_sig <- klf_12d_sig %>% filter(klf14 == 'klf14_related')
klfs_atac_dual_sig <- klf_12d_sig %>% filter(klf3 == 'klf3_related', klf14 == 'klf14_related')

grid.newpage()
png(filename = paste0(figure, 'venn_diagram_atac_motifs_gene.png'))
venn <- draw.pairwise.venn(nrow(klf3_atac_sig), nrow(klf14_atac_sig), nrow(klfs_atac_dual_sig),
                   category = c('KLF3_related', 'KLF14_related'),
                   scaled = TRUE,
                   lty = rep("blank", 2),   # Hide the border of circles.
                   fill = c("light blue", "#bbffbb"),   # Color.
                   cat.pos = c(0, 0),                   # Position of labels.
                   fontfamily = "Arial",                # Font name for numbers.
                   cat.fontface="bold",                 # Font style for labels.
                   cat.fontfamily ="Arial"              # Font name for labels.
)

grid.draw(venn)

dev.off()


write.csv(klf_12d_sig, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/atac_klfs_12d_sig.csv')
write.csv(klf14_atac_sig, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/atac_klf14_12d_sig.csv')

motif_rna <- rnaseq %>%
  na.omit() %>%
  mutate(klf3 = ifelse(TFs %in% klf3$TFs, 'klf3_related', 'klf3_not'),
         klf14 = ifelse(TFs %in% klf14$TFs, 'klf14_related', 'klf14_not'),
         myb = ifelse(TFs %in% myb$TFs, 'myb_related', 'myb_not'),
         DEgenes = ifelse(padj <= 0.05, 'significant', 'not_significant'))


rna_klf3_table <- table('DE gene' = motif_rna$DEgenes, 'klf3 gene' = motif_rna$klf3)
rna_klf3_table

rna_klf3_test <-fisher.test(rna_klf3_table)
rna_klf3_test$estimate
rna_klf3_test$p.value

rna_klf14_table <- table('DE gene' = motif_rna$DEgenes, 'klf14 gene' = motif_rna$klf14)
rna_klf14_table

rna_klf14_test <-fisher.test(rna_klf14_table)
rna_klf14_test$estimate
rna_klf14_test$p.value


rna_myb_table <- table('DE gene' = motif_rna$DEgenes, 'myb gene' = motif_rna$myb)
rna_myb_table

rna_myb_test <-fisher.test(rna_myb_table)
rna_myb_test$estimate
rna_myb_test$p.value


klf_rna_sig <- motif_rna %>% filter(padj < 0.05, (klf3 == 'klf3_related'| klf14 == 'klf14_related')) %>%
  dplyr::select(-myb) %>%
  arrange(log2FoldChange)

klf3_rna_sig <- klf_rna_sig %>% filter(klf3 == 'klf3_related')
klf14_rna_sig <- klf_rna_sig %>% filter(klf14 == 'klf14_related')


write.csv(klf_rna_sig, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/rna_klf_sig.csv')
write.csv(klf14_rna_sig, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/rna_klf14_sig.csv')

grid.newpage()
png(filename = paste0(figure, 'venn_diagram_rna_motifs_gene.png'))
venn <- draw.pairwise.venn(nrow(klf3_rna_sig), nrow(klf14_rna_sig), nrow(klfs_rna_dual_sig),
                   category = c('KLF3_related', 'KLF14_related'),
                   scaled = TRUE,
                   lty = rep("blank", 2),   # Hide the border of circles.
                   fill = c("light blue", "#bbffbb"),   # Color.
                   cat.pos = c(0, 0),                   # Position of labels.
                   fontfamily = "Arial",                # Font name for numbers.
                   cat.fontface="bold",                 # Font style for labels.
                   cat.fontfamily ="Arial"              # Font name for labels.
)

grid.draw(venn)

dev.off()

```



```{r}
atac_12h_up <- read_csv('d:/Projects/UCPH/Work/master_thesis/datasets/go_term_atac_12h_klf14_related_up_regulated.csv')
atac_12h_down <- read_csv('d:/Projects/UCPH/Work/master_thesis/datasets/go_term_atac_12h_klf14_related_down_regulated.csv')

atac_12h_up_bp <- atac_12h_up %>%
  filter(str_detect(source, 'GO:BP')) %>%
  arrange(desc(adjusted_p_value)) %>%
  filter(order(adjusted_p_value) <= 10)

  

atac_12h_up_gene_order <- factor(atac_12h_up_bp$term_name, order = T)

atac_12h_up_plot <- atac_12h_up_bp %>%
  ggplot(aes(x = term_name, y = -log10(adjusted_p_value))) +
  geom_bar(stat = 'identity') +
  coord_flip() +
  # ggtitle('Biological process ') +
  xlab('Biological Process') +
  ylab('-log10 adjusted P value') +
  scale_x_discrete(limits = atac_12h_up_gene_order) +
  theme_bw()
atac_12h_up_plot
ggsave(atac_12h_up_plot, file = paste0(figure, 'go_term_upregulated_atac_12h.png'))
```

```{r}
atac_12h_down_bp <- atac_12h_down %>%
  filter(str_detect(source, 'GO:BP')) %>%
  arrange(desc(adjusted_p_value)) %>%
  filter(order(adjusted_p_value) <= 10)

atac_12h_down_gene_order <- factor(atac_12h_down_bp$term_name, order = T)

atac_12h_down_plot <- atac_12h_down_bp %>%
  ggplot(aes(x = term_name, y = -log10(adjusted_p_value))) +
  geom_bar(stat = 'identity') +
  coord_flip() +
  # ggtitle('Biological process ') +
  xlab('Biological Process') +
  ylab('-log10 adjusted P value') +
  scale_x_discrete(limits = atac_12h_down_gene_order) +
  theme_bw()
atac_12h_down_plot
ggsave(atac_12h_down_plot, file = paste0(figure, 'go_term_downregulated_atac_12h.png'))
```

## rnaseq go_term (intersect with ATACseq)


```{r}

rnaseq <- read.csv(paste0(dataset, 'rnaseq_deseq_final_result_total.csv')) %>%
  na.omit()
atacseq <- read.csv(paste0(dataset, 'result_atac_12h_delete.csv'))%>%
  mutate(symbol = str_split(.$genes, ' ', simplify = T)[,1]) %>%
  dplyr::select(-genes) %>%
  na.omit()

klf14_rna_sig <- read.csv(paste0(dataset, 'rna_klf14_sig.csv'))

inter <- rnaseq %>%
  inner_join(., atacseq, by = 'symbol', suffix = c('_rna', '_atac')) %>%
  filter(padj_atac < 0.05, padj_rna < 0.05) %>%
  dplyr::select(c(gene, merged, symbol, log2FoldChange_rna, padj_rna, log2FoldChange_atac, padj_atac)) %>%
  arrange(desc(log2FoldChange_rna)) %>%
  inner_join(., klf14_rna_sig, by = 'gene', suffix = c('_rna', '_klf14'))

write.csv(inter, file = paste0(dataset, 'intersected_genes_related_klf4.csv'))

```




```{r}
inter_up <- read_csv('d:/Projects/UCPH/Work/master_thesis/datasets/go_term_rnaseq_intersect_atacseq_klf14_related_up_regulated.csv')
inter_down <- read_csv('d:/Projects/UCPH/Work/master_thesis/datasets/go_term_rnaseq_intersect_atacseq_klf14_related_down_regulated.csv')

inter_up_bp <- inter_up %>%
  filter(str_detect(source, 'GO:BP')) %>%
  arrange(desc(adjusted_p_value)) %>%
  filter(order(adjusted_p_value) <= 10)

  

inter_up_gene_order <- factor(inter_up_bp$term_name, order = T)

inter_up_plot <- inter_up_bp %>%
  ggplot(aes(x = term_name, y = -log10(adjusted_p_value))) +
  geom_bar(stat = 'identity') +
  coord_flip() +
  # ggtitle('Biological process ') +
  xlab('Biological Process') +
  ylab('-log10 adjusted P value') +
  scale_x_discrete(limits = inter_up_gene_order) +
  theme_bw()
inter_up_plot
ggsave(inter_up_plot, file = paste0(figure, 'go_term_upregulated_intersect.png'))
```

```{r}
inter_down_bp <- inter_down %>%
  filter(str_detect(source, 'GO:BP')) %>%
  arrange(desc(adjusted_p_value)) %>%
  filter(order(adjusted_p_value) <= 10)

inter_down_gene_order <- factor(inter_down_bp$term_name, order = T)

inter_down_plot <- inter_down_bp %>%
  ggplot(aes(x = term_name, y = -log10(adjusted_p_value))) +
  geom_bar(stat = 'identity') +
  coord_flip() +
  # ggtitle('Biological process ') +
  xlab('Biological Process') +
  ylab('-log10 adjusted P value') +
  scale_x_discrete(limits = inter_down_gene_order) +
  theme_bw()
inter_down_plot
ggsave(inter_down_plot, file = paste0(figure, 'go_term_downregulated_intersect.png'))
```




```{r}
sessionInfo()
```

