---
title: "atacseq_pipeline_hg38_bamfiles"
author: "Sejun Kim"
date: '2021 12 2 '
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
 library(airway)
 library(Rsamtools)
# library(GenomicFeatures)
# library("GenomicAlignments")
 library("BiocParallel")
library(Rsubread)
library(pheatmap)
library(DESeq2)
library(RColorBrewer)
library(factoextra)
library(ggfortify)
library(ggrepel)
library(AnnotationDbi)
library(Homo.sapiens)
library(tximport)
library(stringr)
library(DiffBind)

setwd('d:/Projects/UCPH/Work/atacseq/peak_calling_lower')
register(SnowParam(4))
peak <- 'd:/Projects/UCPH/Work/atacseq/peak_calling_lower/'
scales <- 'd:/Projects/UCPH/Work/atacseq/peak_calling_lower/scale_factor/'
```


# load count data

```{r}

file_list <-list.files(path = peak, pattern = '*_lower.txt')

scale <- data.frame()
data <- data.frame()

datafile <- lapply(paste0(peak, file_list), FUN = read.table, col.names = c('merged', 'counts'))
names(datafile) <- file_list


data_count <- as.data.frame(datafile) %>%
  dplyr::select(ends_with('counts')) %>%
  as.matrix()


rownames(data_count) <- datafile[[1]]$merged

colnames(data_count) <- c('si673_12h_1','si673_12h_2','si673_12h_3',
                                   'si673_24h_1', 'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 'si673_4h_3', 
                     'siNT_1', 'siNT_2', 'siNT_3')

NormFactor <- edgeR::calcNormFactors(object = data_count, method = "TMM")

## if you prefer to use the DESeq2 strategy use method="RLE" instead

## raw library size:
LibSize <- colSums(data_count)

## calculate size factors:
SizeFactors <- NormFactor * LibSize / 1000000

## Reciprocal, please read section below:   
SizeFactors.Reciprocal <- 1/SizeFactors




n <- 1
for (d in SizeFactors.Reciprocal){
  write.table(d, file = paste0(scales, 'scale_factor_',names(datafile)[n]),
            row.names = FALSE, col.names = FALSE)
  n = n + 1
}


# write.table(data_count, file = paste0(peak, 'raw_count.txt'))


```


```{r}
count <- read.table(paste0(peak, 'raw_count.txt')) %>%
  rownames_to_column('merged')
#normal <- read_tsv('Peak_calling/normalized_count.txt')
colnames(count) <- c('merged','si673_12h_1','si673_12h_2','si673_12h_3',
                                   'si673_24h_1', 'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 'si673_4h_3', 
                     'siNT_1', 'siNT_2', 'siNT_3')

anno <- read.delim(paste0(peak, 'peak_annotation.txt'), comment.char = '#', 
                   header = F) %>%
  mutate(merged = V1, 
         gene1 = str_split(str_split(.$V2, ', ', simplify = T)[,1],
                         '\\(', simplify = T)[,1], 
         gene2 = str_split(str_split(.$V2, ', ', simplify = T)[,2],
                         '\\(', simplify = T)[,1], 
         gene3 = str_split(str_split(.$V2, ', ', simplify = T)[,3],
                         '\\(', simplify = T)[,1], 
         gene4 = str_split(str_split(.$V2, ', ', simplify = T)[,4],
                         '\\(', simplify = T)[,1], 
         gene5 = str_split(str_split(.$V2, ', ', simplify = T)[,5],
                         '\\(', simplify = T)[,1], 
         gene6 = str_split(str_split(.$V2, ', ', simplify = T)[,6],
                         '\\(', simplify = T)[,1], 
         gene7 = str_split(str_split(.$V2, ', ', simplify = T)[,7],
                         '\\(', simplify = T)[,1]) %>%
  dplyr::select(-c(V1, V2)) 




anno1 <- anno %>% filter(gene1!="") %>% dplyr::select(merged, genes=gene1)
anno2 <- anno %>% filter(gene2!="") %>% dplyr::select(merged, genes=gene2)
anno3 <- anno %>% filter(gene3!="") %>% dplyr::select(merged, genes=gene3)
anno4 <- anno %>% filter(gene4!="") %>% dplyr::select(merged, genes=gene4)
anno5 <- anno %>% filter(gene5!="") %>% dplyr::select(merged, genes=gene5)
anno6 <- anno %>% filter(gene6!="") %>% dplyr::select(merged, genes=gene6)
anno7 <- anno %>% filter(gene7!="") %>% dplyr::select(merged, genes=gene7)

annot <- bind_rows(anno1, anno2, anno3, anno4, anno5, anno6, anno7) %>% arrange(merged)
# write.table(annot,paste0(peak, 'peak_annotation_2.txt'), row.names = F)

annotation <- annot
# annotation <- read.delim(paste0(peak, 'peak_annotation_2.txt'), header = T,  sep = '\t')

total <- full_join(annotation, count, by = 'merged' ) %>%
  na.omit() 

total_count <- as.matrix(total[3:ncol(total)])
rownames(total_count) <- as.factor(total$genes)



design <- data.frame(condition = c( '12h','12h','12h',
                                   '24h', '24h','24h', '4h', '4h', '4h', 'Ctrl', 'Ctrl', 'Ctrl'))

outlier_delete_design <- data.frame(condition = c( '12h','12h','12h',
                                    '24h','24h', '4h', '4h', 'Ctrl', 'Ctrl', 'Ctrl'))

rownames(outlier_delete_design) <- c('si673_12h_1','si673_12h_2','si673_12h_3',
                                   'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 
                     'siNT_1', 'siNT_2', 'siNT_3')



rownames(design) <- colnames(total)[3:length(total)]
designs <- design %>% t()

ct_vs_12 <- cbind(total_count[,1:3], total_count[,10:12])
ct_vs_24 <- cbind(total_count[,4:6], total_count[,10:12])
ct_vs_4 <- total_count[,7:12]

ct_vs_4_3_delete <- ct_vs_4 %>% as.data.frame() %>% dplyr::select(-si673_4h_3) %>% as.matrix()
ct_vs_24_1_delete <- ct_vs_24 %>% as.data.frame() %>% dplyr::select(-si673_24h_1) %>% as.matrix()

outlier_delete <- total_count %>% as.data.frame() %>% dplyr::select(-c(si673_4h_3, si673_24h_1)) %>% as.matrix()

ct_vs_12_design <- cbind(designs[1], designs[2], designs[3], designs[10], designs[11], designs[12]) %>% t()
rownames(ct_vs_12_design) <- c('si673_12h_1','si673_12h_2','si673_12h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_12_design) <- 'condition'
ct_vs_24_design <- cbind(designs[4], designs[5], designs[6], designs[10], designs[11], designs[12]) %>% t()
rownames(ct_vs_24_design) <- c('si673_24h_1','si673_24h_2','si673_24h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_24_design) <- 'condition'
ct_vs_4_design <- designs[7:12] %>% t() %>% t()
rownames(ct_vs_4_design) <- c('si673_4h_1','si673_4h_2','si673_4h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_4_design) <- 'condition'


ct_vs_4_3_delete_design <- data.frame(ct_vs_4_design[-3])
rownames(ct_vs_4_3_delete_design) <- c('si673_4h_1','si673_4h_2','siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_4_3_delete_design) <- 'condition'

ct_vs_24_1_delete_design <- data.frame(ct_vs_24_design[-1])
rownames(ct_vs_24_1_delete_design) <- c('si673_24h_2','si673_24h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_24_1_delete_design) <- 'condition'



```


# Deseq analysis

```{r}
# DESeq2

dds <- DESeqDataSetFromMatrix(countData = total_count,
                               colData = design,
                               design = ~condition)
deseq <- DESeq(dds)

# sort result with logFC threshold 0, padj < 0.1
result <- results(deseq, contrast = c('condition', '24h', 'Ctrl'))

# confirm result
summary(result)

??TMM

# make tibble from DESeq result
result_tb <- result %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd <- vst(dds) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists <- dist( t( assay(vsd) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap <- pheatmap(sampleDistMatrix, 
                        main = "Sample distance clustering",
                        annotation_col = design)
distheatmap
#ggsave(distheatmap, file = 'sample_dist_heatmap.png')

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor <- cor(assay(vsd))
# generate heatmap
samplecorheatmap <- pheatmap(
  as.matrix(sampleCor),
  main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = design
)
samplecorheatmap
#ggsave(samplecorheatmap, file = 'sample_correlation_heatmap.png')

```


# PCA

```{r}
pca <- prcomp(t(assay(vsd)), center=TRUE, scale.=FALSE)
fviz_eig(pca) # elbow method

pca.x <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var <- summary(pca)$importance['Proportion of Variance',] * 100

pca_2 <- ggplot(pca.x, aes(x = PC1, y = PC2, color = design$condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var[1])) +
  ylab(label = paste('PC2', pca.var[2])) +
  geom_label_repel() +
  theme_bw()

pca_2

ggsave(pca_2, file = 'pca_plot.png')
```


# heatmap of genes for higher padj values


```{r}
x<- assay(vsd)
mat <- x[head(order(result_tb$padj),20), ]

mat <- mat - rowMeans(mat)
adjheatmap <- pheatmap(mat, annotation_col=design, cellheight = 10)
adjheatmap
#ggsave(adjheatmap, file = 'gene_significant_heatmap.png')

```


```{r}
plotDispEsts(deseq)
```



# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tb, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```

# volcano plot

```{r}
# make a plot - volcano plot
result_padj <- result_tb %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

#write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))

subset <- result_padj %>%
  filter(padj < 0.1)

volcano <- ggplot(result_tb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subset, max.overlaps = 1000) +
  theme_bw()
volcano

ggsave(volcano, file = 'volcano_plot_atac_total.png')
```




# Deseq analysis - separately

```{r}
# DESeq2

dds_4 <- DESeqDataSetFromMatrix(countData = ct_vs_4,
                               colData = ct_vs_4_design,
                               design = ~condition)
deseq_4 <- DESeq(dds_4)

# sort result with logFC threshold 0, padj < 0.1
result_4 <- results(deseq_4, contrast = c('condition', '4h', 'Ctrl'))

# confirm result
summary(result_4)

# make tibble from DESeq result
result_tb_4 <- result_4 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot

```{r}
# make a plot - volcano plot
result_padj_4 <- result_tb_4 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

#write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))

subset_4 <- result_padj_4 %>%
  filter(padj < 0.1)

volcano_4 <- ggplot(result_tb_4, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subset_4, max.overlaps = 1000) +
  theme_bw()
volcano_4

ggsave(volcano_4, file = 'volcano_plot_atac_4.png')
```



# Deseq analysis

```{r}
# DESeq2

dds_12 <- DESeqDataSetFromMatrix(countData = ct_vs_12,
                               colData = ct_vs_12_design,
                               design = ~condition)
deseq_12 <- DESeq(dds_12)

# sort result with logFC threshold 0, padj < 0.1
result_12 <- results(deseq_12, contrast = c('condition', '12h', 'Ctrl'))

# confirm result
summary(result_12)

# make tibble from DESeq result
result_tb_12 <- result_12 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot

```{r}
# make a plot - volcano plot
result_padj_12 <- result_tb_12 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

#write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))

subset_12 <- result_padj_12 %>%
  filter(padj < 0.1)

volcano_12 <- ggplot(result_tb_12, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subset_12, max.overlaps = 1000) +
  theme_bw()
volcano_12

ggsave(volcano_12, file = 'volcano_plot_atac_12.png')
```









# Deseq analysis

```{r}
# DESeq2

dds_24 <- DESeqDataSetFromMatrix(countData = ct_vs_24,
                               colData = ct_vs_24_design,
                               design = ~condition)
deseq_24 <- DESeq(dds_24)

# sort result with logFC threshold 0, padj < 0.1
result_24 <- results(deseq_24, contrast = c('condition', '24h', 'Ctrl'))

# confirm result
summary(result_24)

# make tibble from DESeq result
result_tb_24 <- result_24 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_24, file = 'result_atac_24.csv')
```



# volcano plot

```{r}
# make a plot - volcano plot
result_padj_24 <- result_tb_24 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

write_csv(result_padj_24, file = 'high_padj_atac_24.csv')

subset_24 <- result_padj_24 %>%
  filter(padj < 0.1)

volcano_24 <- ggplot(result_tb_24, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subset_24, max.overlaps = 1000) +
  theme_bw()
volcano_24

ggsave(volcano_24, file = 'volcano_plot_atac_24.png')
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_24 <- vst(dds_24) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists_24 <- dist( t( assay(vsd_24) ) )
sampleDistMatrix_24 <- as.matrix( sampleDists_24 ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_24 <- pheatmap(sampleDistMatrix_24, 
                        main = "Sample distance clustering"
                        )
distheatmap_24
ggsave(distheatmap_24, file = 'sample_dist_heatmap_24h.png')

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_24 <- cor(assay(vsd_24))
# generate heatmap
samplecorheatmap_24 <- pheatmap(
  as.matrix(sampleCor_24),
  main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_24
#ggsave(samplecorheatmap, file = 'sample_correlation_heatmap.png')

```


# PCA

```{r}

pca_24 <- prcomp(t(assay(vsd_24)), center=TRUE, scale.=FALSE)
fviz_eig(pca_24) # elbow method

pca.x_24 <- pca_24$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var_24 <- summary(pca_24)$importance['Proportion of Variance',] * 100

pca_2_24 <- ggplot(pca.x_24, aes(x = PC1, y = PC2, color = condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var_24[1])) +
  ylab(label = paste('PC2', pca.var_24[2])) +
  geom_label_repel() +
  theme_bw()

pca_2_24

ggsave(pca_2_24, file = 'pca_plot_24h.png')
```


# heatmap of genes for higher padj values


```{r}
x_24<- assay(vsd_24)
mat_24 <- x_24[head(order(result_tb_24$padj),23), ]

mat_24 <- mat_24 - rowMeans(mat_24)
adjheatmap_24 <- pheatmap(mat_24, cellheight = 10)
adjheatmap_24
ggsave(adjheatmap_24, file = 'gene_significant_heatmap_24h.png')

```


```{r}
plotDispEsts(deseq_24)
```



# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tb_24, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```






# Deseq analysis

```{r}
# DESeq2

dds_24d <- DESeqDataSetFromMatrix(countData = ct_vs_24_1_delete,
                               colData = ct_vs_24_1_delete_design,
                               design = ~condition)
deseq_24d <- DESeq(dds_24d)

# sort result with logFC threshold 0, padj < 0.1
result_24d <- results(deseq_24d, contrast = c('condition', '24h', 'Ctrl'))

# confirm result
summary(result_24d)

# make tibble from DESeq result
result_tb_24d <- result_24d %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_24d, file = 'result_atac_24.csv')
```



# volcano plot

```{r}
# make a plot - volcano plot
result_padj_24d <- result_tb_24d %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

write_csv(result_padj_24d, file = 'high_padj_atac_24.csv')

subset_24d <- result_padj_24d %>%
  filter(padj < 0.1)

volcano_24d <- ggplot(result_tb_24d, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subset_24d, max.overlaps = 1000) +
  theme_bw()
volcano_24d

ggsave(volcano_24d, file = 'volcano_plot_atac_24.png')
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_24d <- vst(dds_24d) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists_24d <- dist( t( assay(vsd_24d) ) )
sampleDistMatrix_24d <- as.matrix( sampleDists_24d ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_24d <- pheatmap(sampleDistMatrix_24d, 
                        main = "Sample distance clustering"
                        )
distheatmap_24d
ggsave(distheatmap_24d, file = 'sample_dist_heatmap_24h.png')

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_24d <- cor(assay(vsd_24d))
# generate heatmap
samplecorheatmap_24d <- pheatmap(
  as.matrix(sampleCor_24d),
  main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_24d
#ggsave(samplecorheatmap, file = 'sample_correlation_heatmap.png')

```


# PCA

```{r}

pca_24d <- prcomp(t(assay(vsd_24d)), center=TRUE, scale.=FALSE)
fviz_eig(pca_24d) # elbow method

pca.x_24d <- pca_24d$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var_24d <- summary(pca_24d)$importance['Proportion of Variance',] * 100

pca_2_24d <- ggplot(pca.x_24d, aes(x = PC1, y = PC2, color = condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var_24d[1])) +
  ylab(label = paste('PC2', pca.var_24d[2])) +
  geom_label_repel() +
  theme_bw()

pca_2_24d

ggsave(pca_2_24d, file = 'pca_plot_24h.png')
```


# heatmap of genes for higher padj values


```{r}
x_24d<- assay(vsd_24d)
mat_24d <- x_24d[head(order(result_tb_24d$padj),23), ]

mat_24d <- mat_24d - rowMeans(mat_24d)
adjheatmap_24d <- pheatmap(mat_24d, cellheight = 10)
adjheatmap_24d
ggsave(adjheatmap_24d, file = 'gene_significant_heatmap_24h.png')

```


```{r}
plotDispEsts(deseq_24d)
```



# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tb_24d, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```



# Deseq analysis

```{r}
# DESeq2

ddsd <- DESeqDataSetFromMatrix(countData = outlier_delete,
                               colData = outlier_delete_design,
                               design = ~condition)
deseqd <- DESeq(ddsd)

# sort result with logFC threshold 0, padj < 0.1
resultd <- results(deseqd, contrast = c('condition', '24h', 'Ctrl'))

# confirm result
summary(resultd)



# make tibble from DESeq result
result_tbd <- resultd %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsdd <- vst(ddsd) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDistsd <- dist( t( assay(vsdd) ) )
sampleDistMatrixd <- as.matrix( sampleDistsd ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmapd <- pheatmap(sampleDistMatrixd, 
                        main = "Sample distance clustering",
                        annotation_col = outlier_delete_design)
distheatmapd
ggsave(distheatmapd, file = 'sample_dist_heatmap_delete.png')

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCord <- cor(assay(vsdd))
# generate heatmap
samplecorheatmapd <- pheatmap(
  as.matrix(sampleCord),
  main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = outlier_delete_design
)
samplecorheatmapd
#ggsave(samplecorheatmap, file = 'sample_correlation_heatmap.png')

```


# PCA

```{r}
pcad <- prcomp(t(assay(vsdd)), center=TRUE, scale.=FALSE)
fviz_eig(pcad) # elbow method

pca.xd <- pcad$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.vard <- summary(pcad)$importance['Proportion of Variance',] * 100

pca_2d <- ggplot(pca.xd, aes(x = PC1, y = PC2, color = outlier_delete_design$condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.vard[1])) +
  ylab(label = paste('PC2', pca.vard[2])) +
  geom_label_repel() +
  theme_bw()

pca_2d

ggsave(pca_2d, file = 'pca_plot_delete.png')
```


# heatmap of genes for higher padj values


```{r}
xd<- assay(vsdd)
matd <- xd[head(order(result_tbd$padj),20), ]

matd <- matd - rowMeans(matd)
adjheatmapd <- pheatmap(matd, annotation_col=outlier_delete_design, cellheight = 10)
adjheatmapd
#ggsave(adjheatmap, file = 'gene_significant_heatmap.png')

```


```{r}
plotDispEsts(deseqd)
```



# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tbd, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```

# volcano plot

```{r}
# make a plot - volcano plot
result_padjd <- result_tbd %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

#write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))

subsetd <- result_padjd %>%
  filter(padj < 0.1)

volcanod <- ggplot(result_tbd, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = genes), data = subsetd, max.overlaps = 1000) +
  theme_bw()
volcanod

ggsave(volcanod, file = 'volcano_plot_atac_total.png')
```





