---
title: "all_data_results"
author: "Sejun Kim"
date: '2022 2 18 '
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
# library(airway)
 library(Rsamtools)
# library(GenomicFeatures)
# library("GenomicAlignments")
 library("BiocParallel")
library(Rsubread)
library(pheatmap)
library(DESeq2)
library(RColorBrewer)
library(factoextra)
library(ggfortify)
library(ggrepel)
library(AnnotationDbi)
library(Homo.sapiens)
library(VennDiagram)

register(SnowParam(8))
figure <- 'd:/Projects/UCPH/Work/master_thesis/figures/'
dataset <- 'd:/Projects/UCPH/Work/master_thesis/datasets/'
atac <- 'd:/Projects/UCPH/Work/atacseq/'
peaks <- paste0(atac,'new/')
motifs <- paste0(peaks, 'motifs/')
rna <- 'd:/Projects/UCPH/Work/rnaseq/'
figures <- 'd:/Projects/UCPH/Work/master_thesis/figures_wo_legend/'

```


## atac deseq


```{r}
count <- read.table('d:/Projects/UCPH/Work/atacseq/new/raw_count_new.txt') %>%
  rownames_to_column('merged')
#normal <- read_tsv('Peak_calling/normalized_count.txt')
colnames(count) <- c('merged','si673_12h_1','si673_12h_2','si673_12h_3',
                                   'si673_24h_1', 'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 'si673_4h_3', 
                     'siNT_1', 'siNT_2', 'siNT_3')

anno <- read.delim('d:/Projects/UCPH/Work/atacseq/new/peak_annotation.txt', comment.char = '#', 
                   header = F) %>%
  mutate(merged = V1, 
         gene1 = str_split(str_split(.$V2, ', ', simplify = T)[,1],
                         '\\(', simplify = T)[,1], 
         gene2 = str_split(str_split(.$V2, ', ', simplify = T)[,2],
                         '\\(', simplify = T)[,1], 
         gene3 = str_split(str_split(.$V2, ', ', simplify = T)[,3],
                         '\\(', simplify = T)[,1], 
         gene4 = str_split(str_split(.$V2, ', ', simplify = T)[,4],
                         '\\(', simplify = T)[,1], 
         gene5 = str_split(str_split(.$V2, ', ', simplify = T)[,5],
                         '\\(', simplify = T)[,1], 
         gene6 = str_split(str_split(.$V2, ', ', simplify = T)[,6],
                         '\\(', simplify = T)[,1], 
         gene7 = str_split(str_split(.$V2, ', ', simplify = T)[,7],
                         '\\(', simplify = T)[,1]) %>%
  dplyr::select(-c(V1, V2)) 




anno1 <- anno %>% filter(gene1!="") %>% dplyr::select(merged, genes=gene1)
anno2 <- anno %>% filter(gene2!="") %>% dplyr::select(merged, genes=gene2)
anno3 <- anno %>% filter(gene3!="") %>% dplyr::select(merged, genes=gene3)
anno4 <- anno %>% filter(gene4!="") %>% dplyr::select(merged, genes=gene4)
anno5 <- anno %>% filter(gene5!="") %>% dplyr::select(merged, genes=gene5)
anno6 <- anno %>% filter(gene6!="") %>% dplyr::select(merged, genes=gene6)
anno7 <- anno %>% filter(gene7!="") %>% dplyr::select(merged, genes=gene7)

annot <- bind_rows(anno1, anno2, anno3, anno4, anno5, anno6, anno7) %>% arrange(merged)
write.table(annot,paste0(dataset, 'atac_peak_annotation_2.txt'), row.names = F)

annotation <- annot %>%
  mutate(genes = str_split(.$genes, ' ', simplify = T)[,1])
# annotation <- read.delim(paste0(peak, 'peak_annotation_2.txt'), header = T,  sep = '\t')

total <- full_join(annotation, count, by = 'merged' ) %>%
  na.omit() 

total_count <- as.matrix(total[3:ncol(total)])




design <- data.frame(treatment = c( 'si673_12h','si673_12h','si673_12h',
                                   'si673_24h', 'si673_24h','si673_24h',
                                   'si673_4h', 'si673_4h', 'si673_4h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)'))

outlier_delete_design <- data.frame(treatment = c( 'si673_12h','si673_12h',
                                   'si673_24h','si673_24h',
                                   'si673_4h', 'si673_4h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)'))

rownames(outlier_delete_design) <- c('si673_12h_1','si673_12h_2',
                                   'si673_24h_2','si673_24h_3', 
                     'si673_4h_1', 'si673_4h_2', 
                     'siNT_1', 'siNT_2', 'siNT_3')



rownames(design) <- colnames(total)[3:length(total)]
designs <- design %>% t()

ct_vs_12 <- cbind(total_count[,1:3], total_count[,10:12])
ct_vs_24 <- cbind(total_count[,4:6], total_count[,10:12])
ct_vs_4 <- total_count[,7:12]

ct_vs_4_3_delete <- ct_vs_4 %>% as.data.frame() %>% dplyr::select(-si673_4h_3) %>% as.matrix()
ct_vs_24_1_delete <- ct_vs_24 %>% as.data.frame() %>% dplyr::select(-si673_24h_1) %>% as.matrix()
ct_vs_12_3_delete <- ct_vs_12 %>% as.data.frame() %>% dplyr::select(-si673_12h_3) %>% as.matrix()

outlier_delete <- total_count %>% as.data.frame() %>% dplyr::select(-c(si673_4h_3, si673_24h_1, si673_12h_3)) %>% as.matrix()

ct_vs_12_design <- cbind(designs[1], designs[2], designs[3], designs[10], designs[11], designs[12]) %>% t() %>% as.data.frame()
rownames(ct_vs_12_design) <- c('si673_12h_1','si673_12h_2','si673_12h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_12_design) <- 'treatment'
ct_vs_24_design <- cbind(designs[4], designs[5], designs[6], designs[10], designs[11], designs[12]) %>% t() %>% as.data.frame()
rownames(ct_vs_24_design) <- c('si673_24h_1','si673_24h_2','si673_24h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_24_design) <- 'treatment'
ct_vs_4_design <- designs[7:12] %>% t() %>% t() %>% as.data.frame()
rownames(ct_vs_4_design) <- c('si673_4h_1','si673_4h_2','si673_4h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_4_design) <- 'treatment'


ct_vs_4_3_delete_design <- c('si673_4h', 'si673_4h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)') %>% as.data.frame()
rownames(ct_vs_4_3_delete_design) <- c('si673_4h_1','si673_4h_2','siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_4_3_delete_design) <- 'treatment'

ct_vs_24_1_delete_design <- c('si673_24h', 'si673_24h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)') %>% as.data.frame()
rownames(ct_vs_24_1_delete_design) <- c('si673_24h_2','si673_24h_3', 'siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_24_1_delete_design) <- 'treatment'

ct_vs_12_3_delete_design <- c('si673_12h', 'si673_12h',
                                   'siNT(Ctrl)', 'siNT(Ctrl)', 'siNT(Ctrl)') %>% as.data.frame()
rownames(ct_vs_12_3_delete_design) <- c('si673_12h_1','si673_12h_2','siNT_1', 'siNT_2', 'siNT_3')
colnames(ct_vs_12_3_delete_design) <- 'treatment'


```


# atac Deseq analysis

```{r}
# DESeq2

dds <- DESeqDataSetFromMatrix(countData = total_count,
                               colData = design,
                               design = ~treatment)
deseq <- DESeq(dds)

# sort result with logFC threshold 0, padj < 0.1
result <- results(deseq, contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))

# confirm result
summary(result)



# make tibble from DESeq result
result_tb <- result %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
write.csv(result_tb,paste0(dataset, 'atac_deseq_total_result.csv'), row.names = F)
```




# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd <- vst(dds) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists <- dist( t( assay(vsd) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap <- pheatmap(sampleDistMatrix, 
                        # main = "Sample distance clustering",
                        annotation_col = design, 
                        annotation_legend = F)
distheatmap
ggsave(distheatmap, file = paste0(figures, 'atac_total_sample_dist_heatmap.png'))

```



# PCA

```{r}
pca <- prcomp(t(assay(vsd)), center=TRUE, scale.=FALSE)


pca.x <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var <- summary(pca)$importance['Proportion of Variance',] * 100

pca_2 <- ggplot(pca.x, aes(x = PC1, y = PC2, color = design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2

ggsave(pca_2, file = paste0(figures,'atac_total_pca_plot.png'))
```

heatmap of genes for higher padj values

# volcano plot

```{r}
# make a plot - volcano plot
result_padj <- result_tb %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

write_csv(result_padj, file = paste0(dataset,'atac_total_deseq_high_padj.csv'))

subset <- result_padj %>%
  filter(padj < 0.1)

volcano <- ggplot(result_tb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.1') +
  geom_text_repel(aes(label = genes), data = subset, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano

ggsave(volcano, file = paste0(figures,'volcano_plot_atac_total.png'))
```




# Deseq analysis - separately 4h

```{r}
# DESeq2

dds_4 <- DESeqDataSetFromMatrix(countData = ct_vs_4,
                               colData = ct_vs_4_design,
                               design = ~treatment)
deseq_4 <- DESeq(dds_4)

# sort result with logFC threshold 0, padj < 0.1
result_4 <- results(deseq_4, contrast = c('treatment', 'si673_4h', 'siNT(Ctrl)'))

# confirm result
summary(result_4)

# make tibble from DESeq result
result_tb_4 <- result_4 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot - 4h

```{r}
# make a plot - volcano plot
result_padj_4 <- result_tb_4 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

write_csv(result_padj, file = paste0(dataset, 'atac_4h_high_padj.csv'))

subset_4 <- result_padj_4 %>%
  filter(padj < 0.1)

volcano_4 <- ggplot(result_tb_4, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.1') +
  geom_text_repel(aes(label = genes), data = subset_4, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_4

ggsave(volcano_4, file = paste0(figures,'volcano_plot_atac_4.png'))
```



# Deseq analysis - 12h

```{r}
# DESeq2

dds_12 <- DESeqDataSetFromMatrix(countData = ct_vs_12,
                               colData = ct_vs_12_design,
                               design = ~treatment)
deseq_12 <- DESeq(dds_12)

# sort result with logFC threshold 0, padj < 0.1
result_12 <- results(deseq_12, contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))

# confirm result
summary(result_12)

# make tibble from DESeq result
result_tb_12 <- result_12 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```



# volcano plot - 12h

```{r}
# make a plot - volcano plot
result_padj_12 <- result_tb_12 %>%
  arrange(padj) %>%
  filter(padj < 0.1) 

write_csv(result_padj, file = paste0(dataset,'atac_12h_high_padj.csv'))

subset_12 <- result_padj_12 %>%
  filter(padj < 0.1)

volcano_12 <- ggplot(result_tb_12, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.1') +
  geom_text_repel(aes(label = genes), data = subset_12, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_12

ggsave(volcano_12, file = paste0(figures,'volcano_plot_atac_12.png'))
```





# Data transfrom - 12h

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_12 <- vst(dds_12) # variance stabilizing transformation

```

# heatmap sample dist - 12h

```{r}

sampleDists_12 <- dist( t( assay(vsd_12) ) )
sampleDistMatrix_12 <- as.matrix( sampleDists_12 ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_12 <- pheatmap(sampleDistMatrix_12
                        # main = "Sample distance clustering"
                        )
distheatmap_12
ggsave(distheatmap_12, file = paste0(figure,'sample_dist_heatmap_12h.png'))

```


# sample correlation plot - 12h

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_12 <- cor(assay(vsd_12))
# generate heatmap
samplecorheatmap_12 <- pheatmap(
  as.matrix(sampleCor_12),
  # main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_12
ggsave(samplecorheatmap_12, file = paste0(figure,'sample_correlation_heatmap_12h.png'))

```


# PCA - 12h

```{r}

pca_12 <- prcomp(t(assay(vsd_12)), center=TRUE, scale.=FALSE)
fviz_eig(pca_12) # elbow method

pca.x_12 <- pca_12$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_12 <- summary(pca_12)$importance['Proportion of Variance',] * 100

pca_2_12 <- ggplot(pca.x_12, aes(x = PC1, y = PC2, color = ct_vs_12_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_12[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_12[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_12

ggsave(pca_2_12, file = paste0(figures,'pca_plot_12h.png'))
```


# heatmap of genes for higher padj values - 12h


```{r, fig.height = 11.5, fig.width = 10}
x_12<- assay(vsd_12)
rownames(x_12) <- result_tb_12$genes
mat_12 <- x_12[head(order(result_tb_12$padj),50), ]

mat_12 <- mat_12 - rowMeans(mat_12)
adjheatmap_12 <- pheatmap(mat_12, cellheight = 10)
adjheatmap_12
ggsave(adjheatmap_12, file = paste0(figure,'gene_significant_heatmap_12h.png'))

```


```{r}
plotDispEsts(deseq_12)
```



# MA plot - 12h

```{r}
# make a plot - MA plot
ggplot(result_tb_12, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```







# Deseq analysis - 24h

```{r}
# DESeq2

dds_24 <- DESeqDataSetFromMatrix(countData = ct_vs_24,
                               colData = ct_vs_24_design,
                               design = ~treatment)
deseq_24 <- DESeq(dds_24)

# sort result with logFC threshold 0, padj < 0.1
result_24 <- results(deseq_24, lfcThreshold = 0.25, alpha = 0.05,
                     contrast = c('treatment', 'si673_24h', 'siNT(Ctrl)'))

# confirm result
summary(result_24)

# make tibble from DESeq result
result_tb_24 <- result_24 %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_24, file = paste0(dataset,'result_atac_24.csv'))
```



# volcano plot - 24h

```{r}
# make a plot - volcano plot
result_padj_24 <- result_tb_24 %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

write_csv(result_padj_24, file = paste0(dataset,'high_padj_atac_24.csv'))

subset_24 <- result_padj_24 %>%
  filter(padj < 0.05)

volcano_24 <- ggplot(result_tb_24, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  geom_text_repel(aes(label = genes), data = subset_24, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_24

ggsave(volcano_24, file = paste0(figures,'volcano_plot_atac_24.png'))
```




# Data transfrom - 24h

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_24 <- vst(dds_24) # variance stabilizing transformation

```

# heatmap sample dist - 24h

```{r}

sampleDists_24 <- dist( t( assay(vsd_24) ) )
sampleDistMatrix_24 <- as.matrix( sampleDists_24 ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_24 <- pheatmap(sampleDistMatrix_24 
                        # main = "Sample distance clustering"
                        )
distheatmap_24
ggsave(distheatmap_24, file = paste0(figure,'sample_dist_heatmap_24h.png'))

```


# sample correlation plot - 24h

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_24 <- cor(assay(vsd_24))
# generate heatmap
samplecorheatmap_24 <- pheatmap(
  as.matrix(sampleCor_24),
  # main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_24
ggsave(samplecorheatmap_24, file = paste0(figure,'sample_correlation_heatmap_24h.png'))

```


# PCA - 24h

```{r}

pca_24 <- prcomp(t(assay(vsd_24)), center=TRUE, scale.=FALSE)
fviz_eig(pca_24) # elbow method

pca.x_24 <- pca_24$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_24 <- summary(pca_24)$importance['Proportion of Variance',] * 100

pca_2_24 <- ggplot(pca.x_24, aes(x = PC1, y = PC2, color = ct_vs_24_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_24[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_24[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_24

ggsave(pca_2_24, file = paste0(figures,'pca_plot_24h.png'))
```


# heatmap of genes for higher padj values - 24h


```{r}
x_24<- assay(vsd_24)
mat_24 <- x_24[head(order(result_tb_24$padj),23), ]

mat_24 <- mat_24 - rowMeans(mat_24)
adjheatmap_24 <- pheatmap(mat_24, cellheight = 10)
adjheatmap_24
ggsave(adjheatmap_24, file = paste0(figure,'gene_significant_heatmap_24h.png'))

```


```{r}
plotDispEsts(deseq_24)
```



# MA plot - 24h

```{r}
# make a plot - MA plot
ggplot(result_tb_24, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```






# Deseq analysis - 24h_delete

```{r}
# DESeq2

dds_24d <- DESeqDataSetFromMatrix(countData = ct_vs_24_1_delete,
                               colData = ct_vs_24_1_delete_design,
                               design = ~treatment)
deseq_24d <- DESeq(dds_24d)

# sort result with logFC threshold 0.25, padj < 0.05
result_24d <- results(deseq_24d,  lfcThreshold = 0.25, alpha = 0.05,
                      contrast = c('treatment', 'si673_24h', 'siNT(Ctrl)'))

# confirm result
summary(result_24d)

# make tibble from DESeq result
result_tb_24d <- result_24d %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_24d, file = paste0(dataset,'result_atac_24_1_delete.csv'))
```



# volcano plot - 24h_delete

```{r}
# make a plot - volcano plot
result_padj_24d <- result_tb_24d %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

write_csv(result_padj_24d, file = paste0(dataset,'high_padj_atac_24h_delete.csv'))

subset_24d <- result_padj_24d %>%
  filter(padj < 0.05)

volcano_24d <- ggplot(result_tb_24d, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  geom_text_repel(aes(label = genes), data = subset_24d, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcano_24d

ggsave(volcano_24d, file = paste0(figures,'volcano_plot_atac_24h_delete.png'))
```




# Data transfrom - 24h_delete

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_24d <- vst(dds_24d) # variance stabilizing transformation

```

# heatmap sample dist - 24h_delete

```{r}

sampleDists_24d <- dist( t( assay(vsd_24d) ) )
sampleDistMatrix_24d <- as.matrix( sampleDists_24d ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_24d <- pheatmap(sampleDistMatrix_24d 
                        # main = "Sample distance clustering"
                        )
distheatmap_24d
ggsave(distheatmap_24d, file = paste0(figure,'sample_dist_heatmap_24h_delete.png'))

```


# sample correlation plot - 24h_delete

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_24d <- cor(assay(vsd_24d))
# generate heatmap
samplecorheatmap_24d <- pheatmap(
  as.matrix(sampleCor_24d),
  # main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_24d
ggsave(samplecorheatmap_24d, file = paste0(figure,'sample_correlation_heatmap_24h_delete.png'))

```


# PCA - 24h_delete

```{r}

pca_24d <- prcomp(t(assay(vsd_24d)), center=TRUE, scale.=FALSE)
fviz_eig(pca_24d) # elbow method

pca.x_24d <- pca_24d$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_24d <- summary(pca_24d)$importance['Proportion of Variance',] * 100

pca_2_24d <- ggplot(pca.x_24d, aes(x = PC1, y = PC2, color = ct_vs_24_1_delete_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_24d[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_24d[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_24d

ggsave(pca_2_24d, file = paste0(figures,'pca_plot_24h_delete.png'))
```


# heatmap of genes for higher padj values - 24h_delete


```{r}
x_24d<- assay(vsd_24d)
mat_24d <- x_24d[head(order(result_tb_24d$padj),23), ]

mat_24d <- mat_24d - rowMeans(mat_24d)
adjheatmap_24d <- pheatmap(mat_24d, cellheight = 10)
adjheatmap_24d
ggsave(adjheatmap_24d, file = paste0(figure,'gene_significant_heatmap_24h_delete.png'))

```


```{r}
plotDispEsts(deseq_24d)
```



# MA plot - 24h_delete

```{r}
# make a plot - MA plot
ggplot(result_tb_24d, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```



# Deseq analysis - total_delete

```{r}
# DESeq2

ddsd <- DESeqDataSetFromMatrix(countData = outlier_delete,
                               colData = outlier_delete_design,
                               design = ~treatment)
deseqd <- DESeq(ddsd)

# sort result with logFC threshold 0, padj < 0.1
resultd <- results(deseqd,  lfcThreshold = 0.25, alpha = 0.05,
                   contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))

# confirm result
summary(resultd)



# make tibble from DESeq result
result_tbd <- resultd %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)
```




# Data transfrom - total_delete

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsdd <- vst(ddsd) # variance stabilizing transformation

```

# heatmap sample dist - total_delete

```{r}

sampleDistsd <- dist( t( assay(vsdd) ) )
sampleDistMatrixd <- as.matrix( sampleDistsd ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmapd <- pheatmap(sampleDistMatrixd, 
                        # main = "Sample distance clustering",
                        annotation_col = outlier_delete_design, 
                        annotation_legend = F)
distheatmapd
ggsave(distheatmapd, file = paste0(figures,'sample_dist_heatmap_total_delete.png'))

```


# sample correlation plot - total_delete

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCord <- cor(assay(vsdd))
# generate heatmap
samplecorheatmapd <- pheatmap(
  as.matrix(sampleCord),
  # main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = outlier_delete_design, 
  annotation_legend = F
)
samplecorheatmapd
ggsave(samplecorheatmapd, file = paste0(figures,'sample_correlation_heatmap_total_delete.png'))

```


# PCA - total_delete

```{r}
pcad <- prcomp(t(assay(vsdd)), center=TRUE, scale.=FALSE)
fviz_eig(pcad) # elbow method

pca.xd <- pcad$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.vard <- summary(pcad)$importance['Proportion of Variance',] * 100

pca_2d <- ggplot(pca.xd, aes(x = PC1, y = PC2, color = outlier_delete_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.vard[1])) +
  ylab(label = paste('PC2', 'with variance', pca.vard[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2d

ggsave(pca_2d, file = paste0(figures,'pca_plot_total_delete.png'))
```


# heatmap of genes for higher padj values - total_delete


```{r}
xd<- assay(vsdd)
matd <- xd[head(order(result_tbd$padj),20), ]

matd <- matd - rowMeans(matd)
adjheatmapd <- pheatmap(matd, annotation_col=outlier_delete_design, 
                        annotation_legend = F, cellheight = 10)
adjheatmapd
ggsave(adjheatmapd, file = paste0(figures,'gene_significant_heatmap_total_delete.png'))

```


```{r}
plotDispEsts(deseqd)
```



# MA plot - total_delete

```{r}
# make a plot - MA plot
ggplot(result_tbd, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```

# volcano plot - total_delete

```{r}
# make a plot - volcano plot
result_padjd <- result_tbd %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

#write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))

subsetd <- result_padjd %>%
  filter(padj < 0.05)

volcanod <- ggplot(result_tbd, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  geom_text_repel(aes(label = genes), data = subsetd, max.overlaps = 1000) +
  theme_bw()+
  theme(legend.position = 'none')
volcanod

ggsave(volcanod, file = paste0(figures,'volcano_plot_atac_total_delete.png'))
```




# Deseq analysis - 12h_delete

```{r}
# DESeq2

dds_12d <- DESeqDataSetFromMatrix(countData = ct_vs_12_3_delete,
                               colData = ct_vs_12_3_delete_design,
                               design = ~treatment)
deseq_12d <- DESeq(dds_12d)

# sort result with logFC threshold 0.25, padj < 0.05
result_12d <- results(deseq_12d, lfcThreshold = 0.25, alpha = 0.05, 
                      contrast = c('treatment', 'si673_12h', 'siNT(Ctrl)'))

# confirm result
summary(result_12d)

# make tibble from DESeq result
result_tb_12d <- result_12d %>%
  as.data.frame() %>%
  rownames_to_column('merged') %>%
  as_tibble() %>%
  mutate(genes = total$genes)

write_csv(result_tb_12d, file = paste0(dataset,'result_atac_12h_delete.csv'))
```



# volcano plot - 12h_delete

```{r}
result_tb_12d <- read.csv(paste0(dataset,'result_atac_12h_delete.csv')) 
# make a plot - volcano plot
result_padj_12d <- result_tb_12d %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

write_csv(result_padj_12d, file = paste0(dataset,'high_padj_atac_12_delete.csv'))

subset_12d <- result_padj_12d %>%
  filter(padj < 0.03, abs(log2FoldChange) > 1.5)

volcano_12d <- ggplot(result_tb_12d, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = c(-0.25, 0.25), alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = -log10(0.05)) +
  xlab('log2 Fold Change') +
  ylab('-log10 adjusted P value') +
  labs(color = 'Adjusted P value < 0.05') +
  # geom_text_repel(aes(label = genes), data = subset_12d, max.overlaps = 1000) +
  theme_bw()
# +
#   theme(legend.position = 'none')
volcano_12d

ggsave(volcano_12d, file = paste0(figure,'volcano_plot_atac_12_delete.png'))
```




# Data transfrom - 12h_delete

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd_12d <- vst(dds_12d) # variance stabilizing transformation

```

# heatmap sample dist - 12h_delete

```{r}

sampleDists_12d <- dist( t( assay(vsd_12d) ) )
sampleDistMatrix_12d <- as.matrix( sampleDists_12d ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap_12d <- pheatmap(sampleDistMatrix_12d
                        # main = "Sample distance clustering"
                        )
distheatmap_12d
ggsave(distheatmap_12d, file = paste0(figure,'sample_dist_heatmap_12h_delete.png'))

```


# sample correlation plot - 12h_delete

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor_12d <- cor(assay(vsd_12d))
# generate heatmap
samplecorheatmap_12d <- pheatmap(
  as.matrix(sampleCor_12d),
  # main = "Correlation clustering",
  width = 12,
  height = 10 )
samplecorheatmap_12d
ggsave(samplecorheatmap_12d, file = paste0(figure,'sample_correlation_heatmap_12h_delete.png'))

```


# PCA - 12h_delete

```{r}

pca_12d <- prcomp(t(assay(vsd_12d)), center=TRUE, scale.=FALSE)
fviz_eig(pca_12d) # elbow method

pca.x_12d <- pca_12d$x %>%
  as.data.frame() %>%
  rownames_to_column('treatment') %>%
  as_tibble()

pca.var_12d <- summary(pca_12d)$importance['Proportion of Variance',] * 100

pca_2_12d <- ggplot(pca.x_12d, aes(x = PC1, y = PC2, color = ct_vs_12_3_delete_design$treatment, 
                           label = treatment)) +
  geom_point() + 
  xlab(label = paste('PC1', 'with variance', pca.var_12d[1])) +
  ylab(label = paste('PC2', 'with variance', pca.var_12d[2])) +
  geom_label_repel() +
  labs(color = 'treatment') +
  theme_bw()+
  theme(legend.position = 'none')

pca_2_12d

ggsave(pca_2_12d, file = paste0(figures,'pca_plot_12h_delete.png'))
```


# heatmap of genes for higher padj values - 12h_delete


```{r, fig.height = 11.5, fig.width = 10}
x_12d<- assay(vsd_12d)
rownames(x_12d) <- result_tb_12d$genes
mat_12d <- x_12d[head(order(result_tb_12d$padj),50), ]

mat_12d <- mat_12d - rowMeans(mat_12d)
adjheatmap_12d <- pheatmap(mat_12d, 
                           annotation_col = ct_vs_12_3_delete_design, 
                           annotation_legend = F,
                           cellheight = 10)
adjheatmap_12d
ggsave(adjheatmap_12d, file = paste0(figures,'gene_significant_heatmap_12h_delete.png'))

```


```{r}
plotDispEsts(deseq_12d)
```



# MA plot - 12h_delete

```{r}
# make a plot - MA plot
ggplot(result_tb_12d, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```


```{r}

sessionInfo()
```


