---
title: "rnaseq_pipeline_hg19"
author: "Sejun Kim"
date: '2021 11 21 '
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(airway)
library(Rsamtools)
library(GenomicFeatures)
library("GenomicAlignments")
 library("BiocParallel")
library(Rsubread)
library(pheatmap)
library(DESeq2)
library(RColorBrewer)
library(factoextra)
library(ggfortify)
library(ggrepel)
library(AnnotationDbi)
library(Homo.sapiens)

setwd('d:/Projects/UCPH/Work/rnaseq/')
register(SnowParam(4))
```


# load the bamfiles

```{r}
dir <- 'bamfile_sejun/'
file_list <- list.files(path = dir, pattern = '^LD0.._hg19.bam')
filenames <- file.path(path = dir, file_list)
bamfiles <- BamFileList(filenames, yieldSize = 2000000)
file.exists(filenames)
seqinfo(bamfiles[1])
```


# make design table

check the condition of each experiment

```{r}
design <- data.frame(condition = c('si_NT', 'si_NT', 'si_LNC1', 'si_LNC1',
                                   'si_LNC2', 'si_LNC2', 'si_NT', 'si_LNC1', 
                                   'si_LNC2'))
```


# make reference 

select hg19, hg38 gtf files. gtf files come from ENSENBL ftp download site

```{r}
gtffile <- file.path('Homo_sapiens.GRCh37.87.gtf.gz')

```


if you want to use GenomicFeatures package, the process below is necessary.

```{r eval=FALSE, include=TRUE}
txdb <- makeTxDbFromGFF(gtffile, format = 'gtf')
ebg <- exonsBy(txdb, by = 'gene')
ebg
```


# Using Genomicfeatures

```{r eval=FALSE, include=TRUE}
se <- summarizeOverlaps(features=ebg,
                        reads=bamfiles,
                        mode="Union",
                        singleEnd=T,
                        ignore.strand=F)


se
assay(se)
count1 <- assay(se)

dds1 <- DESeqDataSetFromMatrix(countData = count1,
                               colData = design,
                               design = ~condition)
deseq1 <- DESeq(dds1)

# sort result with logFC threshold 0.25, padj < 0.05
result1 <- results(deseq1, lfcThreshold = 0.25, alpha = 0.05)

# confirm result
summary(result1)

# make tibble from DESeq result
result_tb1 <- result1 %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  as_tibble()

# make a plot
ggplot(result_tb1, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()


ggplot(result_tb1, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  theme_bw()

vsd1 <- vst(dds1)
class(vsd1)
colData(vsd1)
plotPCA(vsd1, 'condition')

```


# Using Rsubread package

```{r}
fc <- featureCounts(files=filenames,
                    annot.ext=gtffile,
                    isGTFAnnotationFile=TRUE,
                    isPairedEnd=F)

count2 <- fc$counts
```

# DESeq2 object

```{r}
# DESeq2
rownames(design) <- colnames(count2)


dds2 <- DESeqDataSetFromMatrix(countData = count2,
                               colData = design,
                               design = ~condition)
deseq2 <- DESeq(dds2)

# sort result with logFC threshold 0.25, padj < 0.05
result2 <- results(deseq2, lfcThreshold = 0.25, alpha = 0.05)

# confirm result
summary(result2)

# make tibble from DESeq result
result_tb2 <- result2 %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  as_tibble()
```


# annotation result

```{r}
result_tb2$symbol <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb2$entrez <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb2$genename <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

symbol <- result_tb2$symbol
```


# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tb2, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```

# volcano plot

```{r}
# make a plot - volcano plot
result_padj <- result_tb2 %>%
  arrange(padj) %>%
  filter(padj < 0.05) %>%
  filter(-log10(padj) > 250 | abs(log2FoldChange) > 4.0)

write.table(result_padj$symbol, file = 'high_padj_genes.txt')

ggplot(result_tb2, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = symbol), data = result_padj, max.overlaps = 1000) +
  theme_bw()
```

# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd2 <- vst(dds2) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists <- dist( t( assay(vsd2) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
pheatmap(sampleDistMatrix, annotation_col = design)

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor <- cor(assay(vsd2))
# generate heatmap
pheatmap(
  as.matrix(sampleCor),
  main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = design
)
```

# PCA

```{r}

pca <- prcomp(t(assay(vsd2)), center=TRUE, scale.=FALSE)
fviz_eig(pca) # elbow method

pca.x <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var <- summary(pca)$importance['Proportion of Variance',] * 100

pca_2 <- ggplot(pca.x, aes(x = PC1, y = PC2, color = design$condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var[1])) +
  ylab(label = paste('PC2', pca.var[2])) +
  geom_label_repel() +
  theme_bw()

pca_2
```


# heatmap of genes for higher padj values


```{r}
x<- assay(vsd2)
row.names(x) <- symbol
mat <- x[head(order(result_tb2$padj),20), ]

mat <- mat - rowMeans(mat)
pheatmap(mat, annotation_col=design, cellheight = 10)

```



```{r}
sessionInfo()
```



