---
title: "rnaseq_pipeline"
author: "Sejun Kim"
date: '2021 11 21 '
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
# library(airway)
 library(Rsamtools)
# library(GenomicFeatures)
# library("GenomicAlignments")
 library("BiocParallel")
library(Rsubread)
library(pheatmap)
library(DESeq2)
library(RColorBrewer)
#library(factoextra)
library(ggfortify)
library(ggrepel)
library(AnnotationDbi)
library(Homo.sapiens)
library(VennDiagram)

setwd('d:/Projects/UCPH/Work/rnaseq/')
register(SnowParam(8))


```


# load the bamfiles

```{r}
file_storage <- 'd:/big_file_storage/rnaseq/bamfile_sejun/'
genome_storage <- 'd:/big_file_storage/genomes/'
prefix <- 'hg38'
file_list <- list.files(path = file_storage, pattern = 'star.bam$')
filenames <- file.path(path = file_storage, file_list)
bamfiles <- BamFileList(filenames, yieldSize = 2000000)
file.exists(filenames)
seqinfo(bamfiles[1])
```


# make design table

check the condition of each experiment

```{r}
design <- data.frame(condition = c('si_NT', 'si_NT', 'si_LNC1', 'si_LNC1',
                                   'si_LNC2', 'si_LNC2', 'si_NT', 'si_LNC1', 
                                   'si_LNC2'))
```


# make reference 

select hg19, hg38 gtf files. gtf files come from ENSENBL ftp download site

```{r}
gtffile <- file.path(paste0(genome_storage,prefix), 'Homo_sapiens.GRCh38.104.gtf.gz')

```


if you want to use GenomicFeatures package, the process below is necessary.

```{r eval=FALSE, include=TRUE}
txdb <- makeTxDbFromGFF(gtffile, format = 'gtf')
ebg <- exonsBy(txdb, by = 'gene')
ebg
```


# Using Genomicfeatures

```{r eval=FALSE, include=TRUE}
se <- summarizeOverlaps(features=ebg,
                        reads=bamfiles,
                        mode="Union",
                        singleEnd=T,
                        ignore.strand=F)


se
assay(se)
count1 <- assay(se)

dds1 <- DESeqDataSetFromMatrix(countData = count1,
                               colData = design,
                               design = ~condition)
deseq1 <- DESeq(dds1)

# sort result with logFC threshold 0.25, padj < 0.05
result1 <- results(deseq1, lfcThreshold = 0.25, alpha = 0.05)

# confirm result
summary(result1)

# make tibble from DESeq result
result_tb1 <- result1 %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  as_tibble()

# make a plot
ggplot(result_tb1, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()


ggplot(result_tb1, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  theme_bw()

vsd1 <- vst(dds1)
class(vsd1)
colData(vsd1)
plotPCA(vsd1, 'condition')

```


# Using Rsubread package

```{r}
# fc <- featureCounts(files=filenames,
#                     annot.ext=gtffile,
#                     isGTFAnnotationFile=TRUE,
#                     isPairedEnd=F)
# 
# count <- fc$counts
```

# DESeq2 object

```{r}
# DESeq2
count <- read.table('d:/Projects/UCPH/Work/master_thesis/datasets/rnaseq_raw_count.txt')

colnames(count) <- c('LD001', 'LD002', 'LD005', 'LD006', 'LD007', 'LD008', 'LD009', 'LD010', 'LD011')
rownames(design) <- colnames(count)

# write.table(count, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/rnaseq_raw_count.txt')


dds <- DESeqDataSetFromMatrix(countData = count,
                               colData = design,
                               design = ~condition)
deseq <- DESeq(dds)

# sort result with logFC threshold 0.25, padj < 0.05
result <- results(deseq, lfcThreshold = 0.25, alpha = 0.05, 
                  contrast = c('condition', 'si_LNC1', 'si_NT'))

result
  # confirm result
summary(result)

# make tibble from DESeq result
result_tb <- result %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  as_tibble()
```


# annotation result

```{r}
result_tb$symbol <- mapIds(Homo.sapiens,
                     keys=result_tb$gene,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb$entrez <- mapIds(Homo.sapiens,
                     keys=result_tb$gene,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb$genename <- mapIds(Homo.sapiens,
                     keys=result_tb$gene,
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

symbol <- result_tb$symbol
```


# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd <- vst(dds) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists <- dist( t( assay(vsd) ) )
sampleDistMatrix <- as.matrix( sampleDists ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
distheatmap <- pheatmap(sampleDistMatrix, 
                        main = "Sample distance clustering",
                        annotation_col = design)
distheatmap
ggsave(distheatmap, file = 'sample_dist_heatmap.png')

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor <- cor(assay(vsd))
# generate heatmap
samplecorheatmap <- pheatmap(
  as.matrix(sampleCor),
  main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = design
)
samplecorheatmap
ggsave(samplecorheatmap, file = 'sample_correlation_heatmap.png')

```

# PCA

```{r}

pca <- prcomp(t(assay(vsd)), center=TRUE, scale.=FALSE)
fviz_eig(pca) # elbow method

pca.x <- pca$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var <- summary(pca)$importance['Proportion of Variance',] * 100

colors <- c('si_LNC1'='darkgreen',
   'si_LNC2'='red',
   'si_NT'='blue')

pca_2 <- ggplot(pca.x, aes(x = PC1, y = PC2, color = design$condition, 
                           label = design$condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var[1])) +
  ylab(label = paste('PC2', pca.var[2])) +
  geom_label_repel() +
  scale_color_manual(values = colors) +
  theme_bw()

pca_2

ggsave(pca_2, file = 'pca_plot.png')
```


# heatmap of genes for higher padj values


```{r}
x<- assay(vsd)
row.names(x) <- symbol
mat <- x[head(order(result_tb$padj),20), ]

mat <- mat - rowMeans(mat)
adjheatmap <- pheatmap(mat, annotation_col=design, 
                       cellheight = 10)
adjheatmap
ggsave(adjheatmap, file = 'gene_significant_heatmap.png')

```

si-LNC1 is more efficient than si-LNC2. So, the comparison below is done  

# data adjustment

```{r}

# data removing
count2 <- count %>%
  as.data.frame() %>%
  dplyr::select(-c('LD007', 'LD008', 'LD011'))

count2 <- read.table('d:/Projects/UCPH/Work/master_thesis/datasets/rnaseq_raw_count_lnc2_delete.txt')

# write.table(count2, file = 'd:/Projects/UCPH/Work/master_thesis/datasets/rnaseq_raw_count_lnc2_delete.txt')

design2 <- design[c(-5, -6, -9),] %>%
  as.data.frame()

  

# DESeq2
rownames(design2) <- colnames(count2)
colnames(design2) <- 'condition'


dds2 <- DESeqDataSetFromMatrix(countData = count2,
                               colData = design2,
                               design = ~condition)
deseq2 <- DESeq(dds2)

# sort result with logFC threshold 0.25, padj < 0.05
result2 <- results(deseq2, lfcThreshold = 0.25, alpha = 0.05, 
                   contrast = c('condition', 'si_LNC1', 'si_NT'))
result2

# saveRDS(result2, 'result2.rds')
# confirm result
summary(result2)

# make tibble from DESeq result
result_tb2 <- result2 %>%
  as.data.frame() %>%
  rownames_to_column('gene') %>%
  as_tibble()
```


# annotation result_2nd

```{r}
result_tb2$symbol <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="SYMBOL",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb2$entrez <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="ENTREZID",
                     keytype="ENSEMBL",
                     multiVals="first")

result_tb2$genename <- mapIds(Homo.sapiens,
                     keys=result_tb2$gene,
                     column="GENENAME",
                     keytype="ENSEMBL",
                     multiVals="first")

symbol2 <- result_tb2$symbol
write_csv(result_tb2, file = paste0(prefix,'.','result_tb.csv'))

result_padj2 <- result_tb2 %>%
  arrange(padj) %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1) 
write_csv(result_padj2, file = paste0(prefix,'.','result2.csv'))

result_pvalue <- result_tb2 %>%
  filter(pvalue < 0.05) %>%
  arrange(log2FoldChange)
write_csv(result_pvalue, file = paste0(prefix,'.','result_pvalue.csv'))

```


# MA plot

```{r}
# make a plot - MA plot
ggplot(result_tb2, aes(x = baseMean, y = log2FoldChange)) +
  geom_point(aes(color = padj<0.05), alpha = 0.5) +
  scale_x_log10() +
  geom_smooth() +
  geom_hline(yintercept = 0, alpha = 0.75,
             color = 'darkgreen', linetype = 'dashed') +
  theme_bw()
```

# volcano plot

```{r}
# make a plot - volcano plot
result_padj <- result_tb2 %>%
  arrange(padj) %>%
  filter(padj < 0.05) 

write_csv(result_padj, file = paste0(prefix,'.','high_padj.csv'))


subset <- result_padj %>%
  filter(padj < 0.05, abs(log2FoldChange) > 3.0)

volcano <- ggplot(result_tb2, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.05), alpha = 0.5) +
  geom_vline(xintercept = 0, alpha = 0.75,
             color = 'black', linetype = 'dashed') +
  geom_hline(yintercept = 1) +
  geom_text_repel(aes(label = symbol), data = subset, max.overlaps = 1000) +
  theme_bw()
volcano

ggsave(volcano, file = 'volcano_plot.png')
```


# Data transfrom

As a solution, DESeq2 offers transformations for count data that stabilize the variance across the mean: the regularize logarithm (rlog) and the variance stabilizing transformation (VST). These have slightly different implementations, discussed a bit in the DESeq2 paper and in the vignette, but a similar goal of stablizing the variance across the range of values. Both produce log2-like values for high counts. Here we will use the variance stabilizing transformation implemented with the vst function:

```{r}

vsd2 <- vst(dds2) # variance stabilizing transformation

```

# heatmap sample dist

```{r}

sampleDists2 <- dist( t( assay(vsd2) ) )
sampleDistMatrix2 <- as.matrix( sampleDists2 ) 
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255) 
pheatmap(sampleDistMatrix2, annotation_col = design2)

```


# sample correlation plot

```{r}
# calculate correlation between pairs of samples for all possible pairs.
sampleCor2 <- cor(assay(vsd2))
# generate heatmap
pheatmap(
  as.matrix(sampleCor2),
  main = "Correlation clustering",
  width = 12,
  height = 10, 
  annotation_col = design2
)
```

# PCA

```{r}

pca2 <- prcomp(t(assay(vsd2)), center=TRUE, scale.=FALSE)
fviz_eig(pca2) # elbow method

pca.x2 <- pca2$x %>%
  as.data.frame() %>%
  rownames_to_column('condition') %>%
  as_tibble()

pca.var2 <- summary(pca2)$importance['Proportion of Variance',] * 100

pca_22 <- ggplot(pca.x2, aes(x = PC1, y = PC2, color = design2$condition, 
                           label = condition)) +
  geom_point() + 
  xlab(label = paste('PC1', pca.var2[1])) +
  ylab(label = paste('PC2', pca.var2[2])) +
  geom_label_repel() +
  theme_bw()

pca_22
```


# heatmap of genes for higher padj values


```{r, fig.height = 11.5, fig.width = 10}
x2<- assay(vsd2)
row.names(x2) <- symbol2
mat2 <- x2[head(order(result_tb2$padj),50), ]



mat2 <- mat2 - rowMeans(mat2)
adjheatmap2 <- pheatmap(mat2, 
                        annotation_col=design2, 
                        cellheight = 13,
                        fontsize = 13,
                        # cluster_rows = F, 
                        cutree_cols = 2
                        )

adjheatmap2
ggsave(adjheatmap2, file = 'gene_significant_heatmap2.png')
```

#making GCT format

```{r}
counts(dds2)
dds2 <- estimateSizeFactors(dds2)
sizeFactors(dds2)
normal_counts <- counts(dds2, normalized=TRUE)
normal_counts <- as.data.frame(normal_counts)
normal_counts$description <- symbol2
normal_counts <- normal_counts %>% 
  relocate(description) %>% 
  rownames_to_column('gene')
fid <- "rnaseq_for_gsea.gct" 
writeLines(c("#1.2", paste(nrow(normal_counts), ncol(normal_counts) - 2, sep = '\t', collapse="\t")), fid, sep="\n")
write.table(normal_counts, file=fid, quote=FALSE, row.names=FALSE, col.names=TRUE, sep="\t", append = TRUE)
```


```{r}

result_up <- result_tb2 %>%
  filter(padj <0.05, log2FoldChange > 0)

result_down <- result_tb2 %>%
  filter(padj <0.05, log2FoldChange < 0)

grid.newpage()
png(filename = 'venn_diagram_rnaseq_up_down.png')
venn <- draw.pairwise.venn(nrow(result_down), nrow(result_up), 0, 
                   category = c('Downregulated genes', 'Upregulated genes'), 
                   scaled = FALSE, 
                   lty = rep("blank", 2),   # Hide the border of circles.
                   fill = c("light blue", "#bbffbb"),   # Color.
                   cat.pos = c(0, 0),                   # Position of labels.
                   fontfamily = "Arial",                # Font name for numbers.
                   cat.fontface="bold",                 # Font style for labels.
                   cat.fontfamily ="Arial"              # Font name for labels.
)

grid.draw(venn)

dev.off()
```






```{r}
sessionInfo()
```









