{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "incomplete-hospital",
   "metadata": {},
   "source": [
    "# Velocyto flowchart\n",
    "\n",
    "Velocyto -> command(run, run10x, run_smartseq2, run_dropest, dropest_bc_correct) with option (implemented by click package)\n",
    "-> input data(bam/sam/bed....) to run function -> input data to _run(actually this function works) function -> return result from _run -> return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "organizational-officer",
   "metadata": {},
   "outputs": [],
   "source": [
    "#velocyto.py\n",
    "\n",
    "import click # make py file available by line command\n",
    "import logging # tracking\n",
    "import sys\n",
    "from typing import Any # making new data types \n",
    "from collections import OrderedDict # import user_defined dictionary -> preserve the order commands were added.\n",
    "from .run import run\n",
    "from .run10x import run10x\n",
    "from .run_smartseq2 import run_smartseq2\n",
    "from .run_dropest import run_dropest\n",
    "from .dropest_bc_correct import dropest_bc_correct\n",
    "import velocyto._version\n",
    "\n",
    "\n",
    "class NaturalOrderGroup(click.Group):\n",
    "    \"\"\"Command group trying to list subcommands in the order they were added.\n",
    "    Make sure you initialize the `self.commands` with OrderedDict instance.\n",
    "    With decorator, use::\n",
    "        @click.group(cls=NaturalOrderGroup, commands=OrderedDict())\n",
    "    \"\"\"\n",
    "\n",
    "    def list_commands(self, ctx: Any) -> Any:\n",
    "        \"\"\"List command names as they are in commands dict.\n",
    "        If the dict is OrderedDict, it will preserve the order commands\n",
    "        were added.\n",
    "        \"\"\"\n",
    "        return self.commands.keys() # command list\n",
    "\n",
    "\n",
    "@click.version_option(version=velocyto._version.__version__)\n",
    "@click.group(cls=NaturalOrderGroup, commands=OrderedDict(), context_settings=dict(max_content_width=300, terminal_width=300))\n",
    "def cli() -> None:\n",
    "    # NOTE: here is a good place to add a comand to control verbosity/logging-level\n",
    "    logging.basicConfig(stream=sys.stdout, format='%(asctime)s - %(levelname)s - %(message)s', level=logging.DEBUG)\n",
    "    return\n",
    "\n",
    "\n",
    "@click.group(cls=NaturalOrderGroup, commands=OrderedDict(), context_settings=dict(max_content_width=300, terminal_width=300))\n",
    "def tools() -> None:\n",
    "    \"\"\"helper tools for velocyto\n",
    "    \"\"\"\n",
    "    return\n",
    "\n",
    "tools.add_command(dropest_bc_correct)\n",
    "cli.add_command(run)\n",
    "cli.add_command(run10x)\n",
    "cli.add_command(run_dropest)\n",
    "cli.add_command(run_smartseq2)\n",
    "cli.add_command(tools)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "similar-robin",
   "metadata": {},
   "outputs": [],
   "source": [
    "# run.py -> from velocyto command\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "# The glob module finds all the pathnames matching \n",
    "# a specified pattern according to the rules used by the Unix shell, \n",
    "# although results are returned in arbitrary order. \n",
    "import re # regular expression\n",
    "import click\n",
    "import numpy as np\n",
    "import random\n",
    "import string\n",
    "import logging\n",
    "from typing import *\n",
    "import velocyto as vcy\n",
    "from ._run import _run\n",
    "\n",
    "# logging.basicConfig(stream=sys.stdout, format='%(asctime)s - %(levelname)s - %(message)s', level=logging.DEBUG)\n",
    "\n",
    "\n",
    "def id_generator(size: int=6, chars: str=string.ascii_uppercase + string.digits) -> str:\n",
    "    return ''.join(random.choice(chars) for _ in range(size))\n",
    "\n",
    "\n",
    "@click.command(short_help=\"Runs the velocity analysis outputting a loom file\")\n",
    "@click.argument(\"bamfile\", nargs=-1, required=True,\n",
    "                type=click.Path(exists=True,\n",
    "                                file_okay=True,\n",
    "                                dir_okay=False,\n",
    "                                readable=True,\n",
    "                                resolve_path=True))\n",
    "@click.argument(\"gtffile\",\n",
    "                type=click.Path(exists=True,\n",
    "                                file_okay=True,\n",
    "                                dir_okay=False,\n",
    "                                readable=True,\n",
    "                                resolve_path=True))\n",
    "\n",
    "# options\n",
    "@click.option(\"--bcfile\", \"-b\",\n",
    "              help=\"\"\"Valid barcodes file, to filter the bam. If --bcfile is not specified all the cell barcodes will be included.\n",
    "              Cell barcodes should be specified in the bcfile as the `CB` tag for each read\"\"\",\n",
    "              default=None,\n",
    "              show_default=True,\n",
    "              type=click.Path(resolve_path=True,\n",
    "                              file_okay=True,\n",
    "                              dir_okay=False,\n",
    "                              readable=True))\n",
    "@click.option(\"--outputfolder\", \"-o\",\n",
    "              help=\"Output folder, if it does not exist it will be created.\",\n",
    "              default=None,\n",
    "              type=click.Path(exists=False))\n",
    "@click.option(\"--sampleid\", \"-e\",\n",
    "              help=\"The sample name that will be used to retrieve informations from metadatatable\",\n",
    "              default=None,\n",
    "              type=click.Path(exists=False))\n",
    "@click.option(\"--metadatatable\", \"-s\",\n",
    "              help=\"Table containing metadata of the various samples (csv formatted, rows are samples and cols are entries)\",\n",
    "              default=None,\n",
    "              type=click.Path(resolve_path=True,\n",
    "                              file_okay=True,\n",
    "                              dir_okay=False,\n",
    "                              readable=True))\n",
    "@click.option(\"--mask\", \"-m\",\n",
    "              help=\".gtf file containing intervals to mask\",\n",
    "              default=None,\n",
    "              type=click.Path(resolve_path=True,\n",
    "                              file_okay=True,\n",
    "                              dir_okay=False,\n",
    "                              readable=True))\n",
    "@click.option(\"--onefilepercell\", \"-c\",\n",
    "              help=\"\"\"If this flag is used every bamfile passed is interpreted as an independent cell, otherwise multiple files are interpreted as batch of different cells to be analyzed together.\n",
    "              Important: cells reads should not be distributed over multiple bamfiles is not supported!! (default: off)\"\"\",\n",
    "              default=False,\n",
    "              is_flag=True)\n",
    "@click.option(\"--logic\", \"-l\",\n",
    "              help=\"The logic to use for the filtering (default: Default)\",\n",
    "              default=\"Default\")\n",
    "@click.option(\"--without-umi\", \"-U\",\n",
    "              help=\"If this flag is used the data is assumed UMI-less and reads are counted instead of molecules (default: off)\",\n",
    "              default=False,\n",
    "              is_flag=True)\n",
    "@click.option(\"--umi-extension\", \"-u\",\n",
    "              help=\"\"\"In case UMI is too short to guarantee uniqueness (without information from the ampping) set this parameter to `chr`, `Gene` ro `[N]bp`\n",
    "              If set to `chr` the mapping position (binned to 10Gb intervals) will be appended to `UB` (ideal for InDrops+dropEst). If set to `Gene` then the `GX` tag will be appended to the `UB` tag.\n",
    "              If set to `[N]bp` the first N bases of the sequence will be used to extend `UB` (ideal for STRT). (Default: `no`)\"\"\",\n",
    "              default=\"no\")\n",
    "@click.option(\"--multimap\", \"-M\",\n",
    "              help=\"\"\"Consider not unique mappings (not reccomended)\"\"\",\n",
    "              default=False,\n",
    "              is_flag=True)\n",
    "@click.option(\"--samtools-threads\", \"-@\",\n",
    "              help=\"The number of threads to use to sort the bam by cellID file using samtools\",\n",
    "              default=16)\n",
    "@click.option(\"--samtools-memory\",\n",
    "              help=\"The number of MB used for every thread by samtools to sort the bam file\",\n",
    "              default=2048)\n",
    "@click.option(\"--dtype\", \"-t\",\n",
    "              help=\"The dtype of the loom file layers - if more than 6000 molecules/reads per gene per cell are expected set uint32 to avoid truncation (default run: uint32)\",\n",
    "              default=\"uint32\")\n",
    "@click.option(\"--dump\", \"-d\",\n",
    "              help=\"For debugging purposes only: it will dump a molecular mapping report to hdf5. --dump N, saves a cell every N cells. If p is prepended a more complete (but huge) pickle report is printed (default: 0)\",\n",
    "              default=\"0\")\n",
    "@click.option('--verbose', '-v',\n",
    "              help=\"Set the vebosity level: -v (only warnings) -vv (warnings and info) -vvv (warnings, info and debug)\",\n",
    "              count=True, default=1)\n",
    "\n",
    "# velocyto run -> input those files -> input at _run function\n",
    "def run(bamfile: str, gtffile: str,\n",
    "        bcfile: str, outputfolder: str,\n",
    "        sampleid: str, metadatatable: str,\n",
    "        mask: str, onefilepercell: bool, logic: str, without_umi: str, umi_extension: str, multimap: bool,\n",
    "        samtools_threads: int, samtools_memory: int, dtype: str, dump: str, verbose: int,\n",
    "        additional_ca: dict={}) -> None:\n",
    "    \"\"\"Runs the velocity analysis outputting a loom file\n",
    "    BAMFILE bam file with sorted reads\n",
    "    GTFFILE genome annotation file\n",
    "    \"\"\"\n",
    "    return _run(bamfile=bamfile, gtffile=gtffile, bcfile=bcfile, outputfolder=outputfolder,\n",
    "                sampleid=sampleid, metadatatable=metadatatable, repmask=mask, onefilepercell=onefilepercell,\n",
    "                logic=logic, without_umi=without_umi, umi_extension=umi_extension, multimap=multimap, test=False, samtools_threads=samtools_threads,\n",
    "                samtools_memory=samtools_memory, dump=dump, loom_numeric_dtype=dtype, verbose=verbose, additional_ca=additional_ca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "underlying-poster",
   "metadata": {},
   "outputs": [],
   "source": [
    "#run10x.py\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import glob\n",
    "import re\n",
    "import gzip\n",
    "import click\n",
    "import numpy as np\n",
    "import csv\n",
    "from collections import defaultdict\n",
    "import logging\n",
    "from typing import *\n",
    "import velocyto as vcy\n",
    "from ._run import _run\n",
    "\n",
    "\n",
    "logging.basicConfig(stream=sys.stdout, format='%(asctime)s - %(levelname)s - %(message)s', level=logging.DEBUG)\n",
    "\n",
    "\n",
    "@click.command(short_help=\"Runs the velocity analysis for a Chromium Sample\")\n",
    "@click.argument(\"samplefolder\",\n",
    "                type=click.Path(exists=True,\n",
    "                                file_okay=False,\n",
    "                                dir_okay=True,\n",
    "                                readable=True,\n",
    "                                writable=True,\n",
    "                                resolve_path=True))\n",
    "@click.argument(\"gtffile\",\n",
    "                type=click.Path(exists=True,\n",
    "                                file_okay=True,\n",
    "                                dir_okay=False,\n",
    "                                readable=True,\n",
    "                                resolve_path=True))\n",
    "@click.option(\"--metadatatable\", \"-s\",\n",
    "              help=\"Table containing metadata of the various samples (csv fortmated rows are samples and cols are entries)\",\n",
    "              default=None,\n",
    "              type=click.Path(resolve_path=True,\n",
    "                              file_okay=True,\n",
    "                              dir_okay=False,\n",
    "                              readable=True))\n",
    "@click.option(\"--mask\", \"-m\",\n",
    "              help=\".gtf file containing intervals to mask\",\n",
    "              default=None,\n",
    "              type=click.Path(resolve_path=True,\n",
    "                              file_okay=True,\n",
    "                              dir_okay=False,\n",
    "                              readable=True))\n",
    "@click.option(\"--logic\", \"-l\",\n",
    "              help=\"The logic to use for the filtering (default: Default)\",\n",
    "              default=\"Default\")\n",
    "@click.option(\"--multimap\", \"-M\",\n",
    "              help=\"\"\"Consider not unique mappings (not reccomended)\"\"\",\n",
    "              default=False,\n",
    "              is_flag=True)\n",
    "@click.option(\"--samtools-threads\", \"-@\",\n",
    "              help=\"The number of threads to use to sort the bam by cellID file using samtools\",\n",
    "              default=16)\n",
    "@click.option(\"--samtools-memory\",\n",
    "              help=\"The number of MB used for every thread by samtools to sort the bam file\",\n",
    "              default=2048)\n",
    "@click.option(\"--dtype\", \"-t\",\n",
    "              help=\"The dtype of the loom file layers - if more than 6000 molecules/reads per gene per cell are expected set uint32 to avoid truncation (default run_10x: uint16)\",\n",
    "              default=\"uint16\")\n",
    "@click.option(\"--dump\", \"-d\",\n",
    "              help=\"For debugging purposes only: it will dump a molecular mapping report to hdf5. --dump N, saves a cell every N cells. If p is prepended a more complete (but huge) pickle report is printed (default: 0)\",\n",
    "              default=\"0\")\n",
    "@click.option('--verbose', '-v',\n",
    "              help=\"Set the vebosity level: -v (only warinings) -vv (warinings and info) -vvv (warinings, info and debug)\",\n",
    "              count=True, default=1)\n",
    "def run10x(samplefolder: str, gtffile: str,\n",
    "           metadatatable: str, mask: str, logic: str, multimap: bool,\n",
    "           samtools_threads: int, samtools_memory: int, dtype: str, dump: str, verbose: str) -> None:\n",
    "    \"\"\"Runs the velocity analysis for a Chromium 10X Sample\n",
    "    10XSAMPLEFOLDER specifies the cellranger sample folder\n",
    "    GTFFILE genome annotation file\n",
    "    \"\"\"\n",
    "\n",
    "    # Check that the 10X analysis was run successfully\n",
    "    if not os.path.isfile(os.path.join(samplefolder, \"_log\")):\n",
    "        logging.error(\"This is an older version of cellranger, cannot check if the output are ready, make sure of this yourself\")\n",
    "    elif \"Pipestance completed successfully!\" not in open(os.path.join(samplefolder, \"_log\")).read():\n",
    "        logging.error(\"The outputs are not ready\")\n",
    "    bamfile = os.path.join(samplefolder, \"outs\", \"possorted_genome_bam.bam\")\n",
    "\n",
    "    bcmatches = glob.glob(os.path.join(samplefolder, os.path.normcase(\"outs/filtered_gene_bc_matrices/*/barcodes.tsv\")))\n",
    "    if len(bcmatches) == 0:\n",
    "        bcmatches = glob.glob(os.path.join(samplefolder, os.path.normcase(\"outs/filtered_feature_bc_matrix/barcodes.tsv.gz\")))\n",
    "    if len(bcmatches) == 0:\n",
    "        logging.error(\"Can not locate the barcodes.tsv file!\")\n",
    "    bcfile = bcmatches[0]\n",
    "    \n",
    "    outputfolder = os.path.join(samplefolder, \"velocyto\")\n",
    "    sampleid = os.path.basename(samplefolder.rstrip(\"/\").rstrip(\"\\\\\"))\n",
    "    assert not os.path.exists(os.path.join(outputfolder, f\"{sampleid}.loom\")), \"The output already exist. Aborted!\"\n",
    "    additional_ca = {}\n",
    "    try:\n",
    "        tsne_file = os.path.join(samplefolder, \"outs\", \"analysis\", \"tsne\", \"2_components\", \"projection.csv\")\n",
    "        if os.path.exists(tsne_file):\n",
    "            tsne = np.loadtxt(tsne_file, usecols=(1, 2), delimiter=',', skiprows=1)\n",
    "            additional_ca[\"_X\"] = tsne[:, 0].astype('float32')\n",
    "            additional_ca[\"_Y\"] = tsne[:, 1].astype('float32')\n",
    "\n",
    "        clusters_file = os.path.join(samplefolder, \"outs\", \"analysis\", \"clustering\", \"graphclust\", \"clusters.csv\")\n",
    "        if os.path.exists(clusters_file):\n",
    "            labels = np.loadtxt(clusters_file, usecols=(1, ), delimiter=',', skiprows=1)\n",
    "            additional_ca[\"Clusters\"] = labels.astype('int') - 1\n",
    "\n",
    "    except Exception:\n",
    "        logging.error(\"Some IO problem in loading cellranger tsne/pca/kmeans files occurred!\")\n",
    "\n",
    "    return _run(bamfile=(bamfile, ), gtffile=gtffile, bcfile=bcfile, outputfolder=outputfolder,\n",
    "                sampleid=sampleid, metadatatable=metadatatable, repmask=mask, onefilepercell=False,\n",
    "                logic=logic, without_umi=False, umi_extension=\"no\", multimap=multimap, test=False, samtools_threads=samtools_threads,\n",
    "                samtools_memory=samtools_memory, dump=dump, loom_numeric_dtype=dtype, verbose=verbose, additional_ca=additional_ca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suffering-softball",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "terminal-franchise",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "civilian-grain",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "worst-douglas",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "horizontal-respondent",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "worldwide-seeking",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorrect-current",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moderate-framing",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "relative-worker",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "moved-complaint",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "impaired-blend",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fifth-process",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "typical-recording",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "controversial-latitude",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "equipped-planner",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "operating-career",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "banner-humidity",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "animal-adrian",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "specialized-gamma",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "imported-studio",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "communist-witch",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "comprehensive-blind",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
